<html>
    <head>
        <title>RevBayes Manual Pages :: Index Page</title>
        <link rel="stylesheet" type="text/css" href="help.css"/>     

        <script src="jquery.js"></script>
        
        <script>
            function navigateHelp(entry) {
                window.open("pages/" + entry + ".html");
            }
        </script>

        <script>
            var showFunctions, showDistributions, showMoves, showMonitor, showTypes, filter;
            var manualEntries, functionEntries, distributionEntries, moveEntries, monitorEntries, typeEntries;
            var showMore;
            $(document).ready(function() {
                
                // assign variables
                showFunctions = $("#showFunctions");
                showDistributions = $("#showDistributions");
                showMoves = $("#showMoves");
                showMonitors = $("#showMonitors");
                showTypes = $("#showTypes");
                filter = $("#filter");

                manualEntries = $("#manualEntries");
                functionEntries = $("#functionEntries");
                distributionEntries = $("#distributionEntries");
                moveEntries = $("#moveEntries");
                monitorsEntries = $("#monitorsEntries");
                typeEntries = $("#typeEntries");
                
                // hide per default
                manualEntries.find("[id$=less-label]").each(function() {
                    $(this).hide();
                });
                manualEntries.find("[id$=more-content]").each(function() {
                    $(this).hide();
                });

                // configure events
                showFunctions.on('click', function() {
                    functionEntries.slideToggle();
                });
                
                showDistributions.on('click', function() {
                    distributionEntries.slideToggle();
                });
                
                showMoves.on('click', function() {
                    moveEntries.slideToggle();
                });
                
                showMonitors.on('click', function() {
                    monitorEntries.slideToggle();
                });

                showTypes.on('click', function() {
                    typeEntries.slideToggle();
                });

                filter.on('input', function() {
                    doFilter(functionEntries, $(this).val());
                    doFilter(distributionEntries, $(this).val());
                    doFilter(moveEntries, $(this).val());
                    doFilter(monitorEntries, $(this).val());
                    doFilter(typeEntries, $(this).val());
                });

                
            });

            function doFilter(section, filterString) {
                section.children().each(function() {
                    if (filterString.length == 0) {
                        $(this).fadeIn();
                    }
                    else {
                        $(this).hide();
                    }
                });
                if (filterString.length > 0) {
                    section.children("[id^=" + filterString + "]").each(function() {
                        $(this).fadeIn();
                    });
                }
            }

            function showMore(entry) {
                $("#" + entry + "-more-label").toggle();
                $("#" + entry + "-less-label").toggle();

                $("#" + entry + "-more-content").slideDown();
            }

            function showLess(entry) {
                $("#" + entry + "-more-label").toggle();
                $("#" + entry + "-less-label").toggle();
                $("#" + entry + "-more-content").slideUp();
            }

        </script>

    </head>
    <body>
        <div class="filter-bar">
            <div class="filter-controls">
                <label for="showFunctions"> || Show functions</label>
                <input type="checkbox" id="showFunctions" checked="checked" />
                
                <label for="showDistributions"> || Show distributions</label>
                <input type="checkbox" id="showDistributions" checked="checked" />
                
                <label for="showMoves"> || Show moves</label>
                <input type="checkbox" id="showMoves" checked="checked" />
                
                <label for="showMonitors"> || Show monitors</label>
                <input type="checkbox" id="showMonitors" checked="checked" />

                <label for="showTypes"> || Show types</label>
                <input type="checkbox" id="showTypes" checked="checked" />

                <label for="filter"> || Name</label>
                <input type="text" id="filter" placeholder="Filter by name..."/>
            </div>
        </div>
        <br />
        <h1>RevBayes Manual Pages :: Index</h1>

        <div id="manualEntries">
            
            <br/>            
            <div class="section-label"><strong><em>Functions</em></strong></div>            
            <div id="functionEntries"><div id="abs">
    <span onClick="navigateHelp('abs');">
        <a href=""><h2>abs</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The 'abs' function returns the absolute value of a number.
            </p></div>    
    <a href="" onClick="showMore('abs'); return false;" class="indent more-label" id="abs-more-label" >Show more... >></a>
    <a href="" onClick="showLess('abs'); return false;" class="indent more-label" id="abs-less-label" ><< Show less</a>
    <div class="indent more-content" id="abs-more-content"><br />
<h2>Name</h2>
<p class="indent">abs</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The 'abs' function returns the absolute value of a number.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             abs( Real x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The number whose absolute value needs to be computed.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # compute the absolute value of a real number
             number <- -3.0 
             absoluteValueOfTheNumber <- abs(number)
             if (number + absoluteValueOfTheNumber != 0.0) {
             	print("Problem when computing an absolute value.")
             }
             else {
             	print("Correct computation of an absolute value.")
             }             
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ceil'); return false;">ceil</a><br />
<a href="" onClick="navigateHelp('floor'); return false;">floor</a><br />
</p><br />
</div>
</div>

<div id="ceil">
    <span onClick="navigateHelp('ceil');">
        <a href=""><h2>ceil</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The 'ceil' function maps the value of a number to the smallest following integer.
            </p></div>    
    <a href="" onClick="showMore('ceil'); return false;" class="indent more-label" id="ceil-more-label" >Show more... >></a>
    <a href="" onClick="showLess('ceil'); return false;" class="indent more-label" id="ceil-less-label" ><< Show less</a>
    <div class="indent more-content" id="ceil-more-content"><br />
<h2>Name</h2>
<p class="indent">ceil</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The 'ceil' function maps the value of a number to the smallest following integer.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             ceil( RealPos x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The number whose ceiling value needs to be obtained.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # compute the ceiling value of a real number
             number <- 3.4 
             ceil(number)         
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('floor'); return false;">floor</a><br />
<a href="" onClick="navigateHelp('abs'); return false;">abs</a><br />
</p><br />
</div>
</div>

<div id="clade">
    <span onClick="navigateHelp('clade');">
        <a href=""><h2>clade</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
                The 'clade' function builds a clade with n leaves from a vector of n strings. 
                The strings become the leaf names. 
                The clade is multifurcated if there are more than 2 leaves.
            </p></div>    
    <a href="" onClick="showMore('clade'); return false;" class="indent more-label" id="clade-more-label" >Show more... >></a>
    <a href="" onClick="showLess('clade'); return false;" class="indent more-label" id="clade-less-label" ><< Show less</a>
    <div class="indent more-content" id="clade-more-content"><br />
<h2>Name</h2>
<p class="indent">clade</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                The 'clade' function builds a clade with n leaves from a vector of n strings. 
                The strings become the leaf names. 
                The clade is multifurcated if there are more than 2 leaves.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             clade( String[] taxa )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >taxa</b>
<div class="indent">
A vector of strings representing leaf names.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String[]<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # Creating a clade from a string
			 myClade <- clade("A")
			 print( myClade )
			 myClade2 <- clade(["A","B","C"])
			 print( myClade2 )
			 type(myClade2)
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
</div>
</div>

<div id="clear">
    <span onClick="navigateHelp('clear');">
        <a href=""><h2>clear</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Clear the current workspace.
            </p></div>    
    <a href="" onClick="showMore('clear'); return false;" class="indent more-label" id="clear-more-label" >Show more... >></a>
    <a href="" onClick="showLess('clear'); return false;" class="indent more-label" id="clear-less-label" ><< Show less</a>
    <div class="indent more-content" id="clear-more-content"><br />
<h2>Name</h2>
<p class="indent">clear</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Clear the current workspace.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             clear( RevObject x = NULL )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
This argument will be removed.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RevObject<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            The clear function removes either a given variable or all variables from the workspace.
            Clearing the workspace is very useful between analysis if you do not want to have
            old connections between variables hanging around.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
             ls()   # check what is in the workspace
             a <- 1
             b := exp(a)
             ls()   # check what is in the workspace
             clear()
             ls()   # check what is in the workspace
             a <- 1
             b := exp(a)
             ls()   # check what is in the workspace
             clear( b )
             ls()   # check what is in the workspace
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ls'); return false;">ls</a><br />
</p><br />
</div>
</div>

<div id="exists">
    <span onClick="navigateHelp('exists');">
        <a href=""><h2>exists</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'exists' determines whether the RevBayes workspace contains a variable named 'name'
            </p></div>    
    <a href="" onClick="showMore('exists'); return false;" class="indent more-label" id="exists-more-label" >Show more... >></a>
    <a href="" onClick="showLess('exists'); return false;" class="indent more-label" id="exists-less-label" ><< Show less</a>
    <div class="indent more-content" id="exists-more-content"><br />
<h2>Name</h2>
<p class="indent">exists</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'exists' determines whether the RevBayes workspace contains a variable named 'name'
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             exists( String name )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >name</b>
<div class="indent">
The variable name<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            'exists' returns 'true' if the workspace contains a variable whose name matches the String 'name' and 'false' otherwise. One use of 'exists' is to add Move and Monitor objects conditional on the variable 'x' existing. The function 'ls' provides a summary for all variable names that 'exists' would evaluate as 'true'.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > ## Correct usage: does "x" exist?
            > x <- 1.0
            > exists("x")
                true
            >
            > ## Incorrect usage: does "1.0" exist?
            > exists(x)
                false
            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ls'); return false;">ls</a><br />
</p><br />
</div>
</div>

<div id="exp">
    <span onClick="navigateHelp('exp');">
        <a href=""><h2>exp</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The 'exp' function returns the exponential value of a number.
            </p></div>    
    <a href="" onClick="showMore('exp'); return false;" class="indent more-label" id="exp-more-label" >Show more... >></a>
    <a href="" onClick="showLess('exp'); return false;" class="indent more-label" id="exp-less-label" ><< Show less</a>
    <div class="indent more-content" id="exp-more-content"><br />
<h2>Name</h2>
<p class="indent">exp</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The 'exp' function returns the exponential value of a number.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             exp( Real x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The number whose exponential needs to be computed.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # Let's compute the exponential of a few numbers
             exp(-1)
             exp(0)
             exp(1)
             exp(10)
             # ln is the exponential's inverse function
             ln(exp(10))
             # Another way of achieving the same thing:
             log(exp(10),base=exp(1))
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ln'); return false;">ln</a><br />
<a href="" onClick="navigateHelp('log'); return false;">log</a><br />
<a href="" onClick="navigateHelp('power'); return false;">power</a><br />
</p><br />
</div>
</div>

<div id="floor">
    <span onClick="navigateHelp('floor');">
        <a href=""><h2>floor</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The 'floor' function maps the value of a number to the largest preceding integer.
            </p></div>    
    <a href="" onClick="showMore('floor'); return false;" class="indent more-label" id="floor-more-label" >Show more... >></a>
    <a href="" onClick="showLess('floor'); return false;" class="indent more-label" id="floor-less-label" ><< Show less</a>
    <div class="indent more-content" id="floor-more-content"><br />
<h2>Name</h2>
<p class="indent">floor</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The 'floor' function maps the value of a number to the largest preceding integer.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             floor( RealPos x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The number whose floor value needs to be obtained.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # compute the floor value of a real number
             number <- 3.4 
             floor(number)         
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ceil'); return false;">ceil</a><br />
<a href="" onClick="navigateHelp('abs'); return false;">abs</a><br />
</p><br />
</div>
</div>

<div id="fnDPPConcFromMean">
    <span onClick="navigateHelp('fnDPPConcFromMean');">
        <a href=""><h2>fnDPPConcFromMean</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This function calculates the approximate expected value for the Dirichlet process prior distribution concentration parameter given a mean number of parameter categories for a specified number of elements.
            </p></div>    
    <a href="" onClick="showMore('fnDPPConcFromMean'); return false;" class="indent more-label" id="fnDPPConcFromMean-more-label" >Show more... >></a>
    <a href="" onClick="showLess('fnDPPConcFromMean'); return false;" class="indent more-label" id="fnDPPConcFromMean-less-label" ><< Show less</a>
    <div class="indent more-content" id="fnDPPConcFromMean-more-content"><br />
<h2>Name</h2>
<p class="indent">fnDPPConcFromMean</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This function calculates the approximate expected value for the Dirichlet process prior distribution concentration parameter given a mean number of parameter categories for a specified number of elements.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             fnDPPConcFromMean( RealPos num_cats, RealPos num_elements )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >num_cats</b>
<div class="indent">
The mean number of rate categories.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >num_elements</b>
<div class="indent">
The number of elements in the DPP distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            This function computes an approximation of the expected value of the concentration parameter (a) for 
            a given number of elements (N) and a given number of parameter categories (k) based on the formula given
            in Liu (1996). (See also McAuliffe et al. 2006, Jara et al. 2007, and Heath et al. 2012.)
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
	 ## number of DPP elements
	 N <- 10
 
	 ## prior mean number of categories
	 k <- 8
 
	 ## concentration parameter of the DPP
	 a <- fnDPPConcFromMean(k, N)
 
	 ## G_0 distribution
	 g_0 = dnExponential(1.0)
 
	 ## DPP distribution constructior
	 rates ~ dnDPP(a, g_0, N)
             </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Liu JS. 1996. Nonparametric hierarchical Bayes via sequential imputa- tions. Ann Stat. 3:911–930.
            <br />
<br />
<br />

                McAuliffe JD, Blei DM, Jordan MI. 2006. Nonparametric empirical Bayes for the Dirichlet process mixture model. Stat Comput. 16: 5–14.
            <br />
<br />
<br />

            	Jara A, Garcia-Zattera MJ, Lesaffre E. 2007. A Dirichlet process mixture model for the analysis of correlated binary responses. Comput Stat Data Anal. 51:5402–5415.
            <br />
<br />
<br />

            	Heath TA, Holder MT, Huelsenbeck JP. 2012. A Dirichlet process prior for estimating lineage-specific substitution rates. Mol. Biol. Evol. 29(3):939–955. 
            <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Tracy A. Heath</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('fnDPPMeanFromConc'); return false;">fnDPPMeanFromConc</a><br />
<a href="" onClick="navigateHelp('dnDPP'); return false;">dnDPP</a><br />
</p><br />
</div>
</div>

<div id="fnDPPMeanFromConc">
    <span onClick="navigateHelp('fnDPPMeanFromConc');">
        <a href=""><h2>fnDPPMeanFromConc</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This function calculates the expected number of parameter categories for a given concentration parameter
            and number of elements under the Dirichlet process prior.
            </p></div>    
    <a href="" onClick="showMore('fnDPPMeanFromConc'); return false;" class="indent more-label" id="fnDPPMeanFromConc-more-label" >Show more... >></a>
    <a href="" onClick="showLess('fnDPPMeanFromConc'); return false;" class="indent more-label" id="fnDPPMeanFromConc-less-label" ><< Show less</a>
    <div class="indent more-content" id="fnDPPMeanFromConc-more-content"><br />
<h2>Name</h2>
<p class="indent">fnDPPMeanFromConc</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This function calculates the expected number of parameter categories for a given concentration parameter
            and number of elements under the Dirichlet process prior.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             fnDPPMeanFromConc( RealPos concentration, RealPos num_elements )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >concentration</b>
<div class="indent">
The concentration parameter of the DPP.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >num_elements</b>
<div class="indent">
The number of elements.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            This function computes the expected number of parameter categories (k) for
            a given number of elements (N) and a given concentration parameter (a) based on the formula given
            in Liu (1996). (See also McAuliffe et al. 2006, Jara et al. 2007, and Heath et al. 2012.)
            </p>
<p class="indent">
            E(k | a, N) = ∑[ a / (a + i - 1) ]
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
	 ## number of DPP elements
	 N <- 10
 
	 ## prior mean number of categories
	 a <- 3.3015
 
	 ## concentration parameter of the DPP
	 k <- fnDPPMeanFromConc(a, N)
 
	 ## G_0 distribution
	 g_0 = dnExponential(1.0)
 
	 ## DPP distribution constructior
	 rates ~ dnDPP(a, g_0, N)
             </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Liu JS. 1996. Nonparametric hierarchical Bayes via sequential imputa- tions. Ann Stat. 3:911–930.
            <br />
<br />
<br />

                McAuliffe JD, Blei DM, Jordan MI. 2006. Nonparametric empirical Bayes for the Dirichlet process mixture model. Stat Comput. 16: 5–14.
            <br />
<br />
<br />

            	Jara A, Garcia-Zattera MJ, Lesaffre E. 2007. A Dirichlet process mixture model for the analysis of correlated binary responses. Comput Stat Data Anal. 51:5402–5415.
            <br />
<br />
<br />

            	Heath TA, Holder MT, Huelsenbeck JP. 2012. A Dirichlet process prior for estimating lineage-specific substitution rates. Mol. Biol. Evol. 29(3):939–955. 
            <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Tracy A. Heath</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('fnDPPConcFromMean'); return false;">fnDPPConcFromMean</a><br />
<a href="" onClick="navigateHelp('dnDPP'); return false;">dnDPP</a><br />
</p><br />
</div>
</div>

<div id="fnDiscretizeGamma">
    <span onClick="navigateHelp('fnDiscretizeGamma');">
        <a href=""><h2>fnDiscretizeGamma</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The discretized gamma distribution gives the quantiles of the gamma distribution.
            </p></div>    
    <a href="" onClick="showMore('fnDiscretizeGamma'); return false;" class="indent more-label" id="fnDiscretizeGamma-more-label" >Show more... >></a>
    <a href="" onClick="showLess('fnDiscretizeGamma'); return false;" class="indent more-label" id="fnDiscretizeGamma-less-label" ><< Show less</a>
    <div class="indent more-content" id="fnDiscretizeGamma-more-content"><br />
<h2>Name</h2>
<p class="indent">fnDiscretizeGamma</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The discretized gamma distribution gives the quantiles of the gamma distribution.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             fnDiscretizeGamma( RealPos shape, RealPos rate, Integer num_cats, Bool median = FALSE )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >shape</b>
<div class="indent">
The shape parameter of the gamma distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >rate</b>
<div class="indent">
The rate parameter of the gamma distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >num_cats</b>
<div class="indent">
Number of categories in which to break the distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Integer<br />
</div>
<br />
<b >median</b>
<div class="indent">
Should we use the median or the mean within each quantile.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
FALSE<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            In some cases you want to use the quantiles of a distribution instead of estimating the continuous value of a distribution.
            You could compute the k quantiles yourself. This function, however, provides you a convenient access to compute the gamma quantiles (discretized gamma)
            because it is so often used for modeling rate variation among sites.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                s <- 2
                r <- 2
                a := fnDiscretizeGamma(shape=s,rate=r,num_cats=4)
                a
                mean(a)
                s <- 4
                a
                mean(a)
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnGamma'); return false;">dnGamma</a><br />
</p><br />
</div>
</div>

<div id="fnGTR">
    <span onClick="navigateHelp('fnGTR');">
        <a href=""><h2>fnGTR</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The function fnGTR creates the most general reversible rate matrix on an  alphabet with an arbitrary number of states.
            </p></div>    
    <a href="" onClick="showMore('fnGTR'); return false;" class="indent more-label" id="fnGTR-more-label" >Show more... >></a>
    <a href="" onClick="showLess('fnGTR'); return false;" class="indent more-label" id="fnGTR-less-label" ><< Show less</a>
    <div class="indent more-content" id="fnGTR-more-content"><br />
<h2>Name</h2>
<p class="indent">fnGTR</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The function fnGTR creates the most general reversible rate matrix on an  alphabet with an arbitrary number of states.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             fnGTR( Simplex exchangeRates, Simplex baseFrequencies )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >exchangeRates</b>
<div class="indent">
A simplex giving the exchange rate parameters. Size 6 for an alphabet of size 4, like DNA.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
</div>
<br />
<b >baseFrequencies</b>
<div class="indent">
A simplex giving the equilibrium state frequencies. Size the size of the alphabet (4 for DNA).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            The General Time Reversible (GTR) model is the most general reversible Markov model possible. More complex models are not reversible (for a discussion of reversibility in models of sequence evolution, see e.g. Jayaswal et al., 2010). All other reversible models can be seed as specific cases of the GTR model. It has been first proposed by Lanave et al. (1984), Tavaré et al. (1986), and Rodriguez et al. (1990), and is sometimes called "REV" (Yang, 1994).
            </p>
<p class="indent">
            It is usually specified thanks to two sets of parameters, exchangeability rates (a simplex with 6 values for alphabets of size 4 like DNA) and equilibrium frequencies (a simplex with 4 values for alphabets of size 4 like DNA). If sequences are evolved for an infinite amount of time using a GTR matrix, their state frequencies will converge to the equilibrium frequencies of the GTR matrix. In total, a GTR matrix for DNA has 8 free parameters: 5 independent exchangeability rates (the sixth one can be obtained from the other ones, because they sum to 1.0), and 3 independent equilibrium frequencies (the frequencies also sum to one).
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
             # Let's simulate data using a GTR model on a DNA alphabet.
             er_prior <- v(1,1,1,1,1,1)
  			 er ~ dnDirichlet(er_prior)

  			 sf_prior <- v(1,1,1,1)
  			 sf ~ dnDirichlet(sf_prior)

  			 matrix := fnGTR(er,sf) 
             print(matrix)
             
             # Let's change the matrix so that changes towards "G" are much more frequent than others. To achieve that, we put a high "G" equilibrium frequency.
             # We create a simplex:
             GBiasedFrequencies <- simplex(0.01, 0.01, 0.97, 0.01)
             sf.setValue(GBiasedFrequencies)
             
             #What does the matrix look like now?
             print(matrix)
             
             # We build a tree with 3 leaves
             n_species <- 3
             for (i in 1:n_species) {
        			s_names[i] = "Species_"+i
			 }
             speciation ~ dnExponential(10.0)
			 extinction ~ dnExponential(10.0)
			 tree_height ~ dnUnif(0,1.0)
			 speciesTree ~ dnBDP(lambda=speciation, mu=extinction, origin=tree_height, nTaxa=n_species, names=s_names)
             
             # Let's simulate 30 sites over this tree, starting from a sequence completely made of "A"s.
             # Root frequencies: all "A"s
             frequencyA <- 1.0
             rootFrequencies <- simplex(frequencyFirstState, 0.0, 0.0, 0.0)
             n_sites=30
             aln ~ dnPhyloCTMC(tree=speciesTree, Q=matrix, rootFrequencies=rootFrequencies, nSites=n_sites, type="DNA")
			
			 # Now we can have a look at the sequences in the resulting alignment. If things worked well, we should see traces of a vast increase in "G"s from an ancestral sequence entirely made of "A"s.
			 aln.show()             
             
             # And we can compute the log probability of the alignment
             aln.lnProbability()
             </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
        	Jayaswal V, Jermiin L, Pladian L, Robinson J. 2010. Two Stationary Nonhomogeneous Markov Models of Nucleotide Sequence Evolution, Syst. Biology, 60, 74-86.
            <br />
<br />
10.1093/sysbio/syq076<br />
</p><br />
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('fnJC'); return false;">fnJC</a><br />
<a href="" onClick="navigateHelp('fnHKY'); return false;">fnHKY</a><br />
<a href="" onClick="navigateHelp('fnF81'); return false;">fnF81</a><br />
<a href="" onClick="navigateHelp('fnPomo'); return false;">fnPomo</a><br />
</p><br />
</div>
</div>

<div id="fnHKY">
    <span onClick="navigateHelp('fnHKY');">
        <a href=""><h2>fnHKY</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The function fnHKY creates the reversible DNA rate matrix proposed by Hasegawa, Kishino and Yano in 1985.
            </p></div>    
    <a href="" onClick="showMore('fnHKY'); return false;" class="indent more-label" id="fnHKY-more-label" >Show more... >></a>
    <a href="" onClick="showLess('fnHKY'); return false;" class="indent more-label" id="fnHKY-less-label" ><< Show less</a>
    <div class="indent more-content" id="fnHKY-more-content"><br />
<h2>Name</h2>
<p class="indent">fnHKY</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The function fnHKY creates the reversible DNA rate matrix proposed by Hasegawa, Kishino and Yano in 1985.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             fnHKY( RealPos kappa, Simplex baseFrequencies )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >kappa</b>
<div class="indent">
The transition/transversion ratio.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >baseFrequencies</b>
<div class="indent">
A simplex giving the equilibrium base frequencies.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            The HKY rate matrix is defined by 5 parameters: the transition/transversion ratio kappa, and the 4 equilibrium base frequencies. In total, there are 4 independent parameters because the base frequencies sum to 1.  
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
             # an executable example of the function described here
             trueTree <- readTrees("data/primates.tree")[1]
             ...
             </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Hasegawa M, Kishino H and Yano T. 1985. Dating of the human-ape splitting by a molecular clock of mitochondrial DNA. Molecular Biology And Evolution, 22(2), 160-174. 
            <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Your name here</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('fnJC'); return false;">fnJC</a><br />
<a href="" onClick="navigateHelp('fnGTR'); return false;">fnGTR</a><br />
<a href="" onClick="navigateHelp('fnF81'); return false;">fnF81</a><br />
<a href="" onClick="navigateHelp('fnPomo'); return false;">fnPomo</a><br />
</p><br />
</div>
</div>

<div id="fnJC">
    <span onClick="navigateHelp('fnJC');">
        <a href=""><h2>fnJC</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The function fnJC creates the simplest rate matrix on an alphabet with a given number of states.
            </p></div>    
    <a href="" onClick="showMore('fnJC'); return false;" class="indent more-label" id="fnJC-more-label" >Show more... >></a>
    <a href="" onClick="showLess('fnJC'); return false;" class="indent more-label" id="fnJC-less-label" ><< Show less</a>
    <div class="indent more-content" id="fnJC-more-content"><br />
<h2>Name</h2>
<p class="indent">fnJC</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The function fnJC creates the simplest rate matrix on an alphabet with a given number of states.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             fnJC( Natural numStates )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >numStates</b>
<div class="indent">
the number of states in the alphabet.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            The Jukes and Cantor model (1969) is the simplest substitution model and requires no parameter. It assumes equal transition rates among any two states. It works on alphabets of any size but was initially described by Jukes and Cantor for DNA (4 states). It is fast but makes a number of usually unrealistic assumptions, including equal base frequencies in the sequences, and equal mutation rates.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
             # Let's simulate data using a Jukes and Cantor model on an alphabet with 5 states.
             matrix <- fnJC(numStates=5)
             print(matrix)
             
             # We build a tree with 3 leaves
             n_species <- 3
             for (i in 1:n_species) {
        			s_names[i] = "Species_"+i
			 }
             speciation ~ dnExponential(10.0)
			 extinction ~ dnExponential(10.0)
			 tree_height ~ dnUnif(0,1.0)
			 speciesTree ~ dnBDP(lambda=speciation, mu=extinction, origin=tree_height, nTaxa=n_species, names=s_names)
             
             # Let's simulate 30 sites over this tree, starting from a sequence all in the first state.
             # Root frequencies: all in the first state
             frequencyFirstState <- 1.0
             rootFrequencies <- simplex(frequencyFirstState, 0.0, 0.0, 0.0, 0.0)
             n_sites=30
             aln ~ dnPhyloCTMC(tree=speciesTree, Q=matrix, rootFrequencies=rootFrequencies, nSites=n_sites, type="Standard")
			
			 # Now we can have a look at the sequences in the resulting alignment
			 aln.show()             
             
             # And we can compute the log probability of the alignment
             aln.lnProbability()
             </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                T.H. Jukes, C.R. Cantor. 1969. Evolution of protein molecules. 
In Mammalian protein metabolism, Vol. III, pp. 21-132.
            <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('fnGTR'); return false;">fnGTR</a><br />
<a href="" onClick="navigateHelp('fnHKY'); return false;">fnHKY</a><br />
<a href="" onClick="navigateHelp('fnF81'); return false;">fnF81</a><br />
<a href="" onClick="navigateHelp('fnPomo'); return false;">fnPomo</a><br />
</p><br />
</div>
</div>

<div id="ln">
    <span onClick="navigateHelp('ln');">
        <a href=""><h2>ln</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The "ln" function computes the natural logarithm of a number.
            </p></div>    
    <a href="" onClick="showMore('ln'); return false;" class="indent more-label" id="ln-more-label" >Show more... >></a>
    <a href="" onClick="showLess('ln'); return false;" class="indent more-label" id="ln-less-label" ><< Show less</a>
    <div class="indent more-content" id="ln-more-content"><br />
<h2>Name</h2>
<p class="indent">ln</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The "ln" function computes the natural logarithm of a number.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             ln( RealPos x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The number whose natural logarithm needs to be computed.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # Let's compute the natural logarithm of a few numbers
             ln(1)
             ln(10)
             ln(exp(1))
             # Let's see what happens when we go beyond the definition range of ln
             ln(-1)
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('log'); return false;">log</a><br />
<a href="" onClick="navigateHelp('exp'); return false;">exp</a><br />
<a href="" onClick="navigateHelp('power'); return false;">power</a><br />
</p><br />
</div>
</div>

<div id="log">
    <span onClick="navigateHelp('log');">
        <a href=""><h2>log</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The "log" function computes the logarithm of a number, given a particular base.
            </p></div>    
    <a href="" onClick="showMore('log'); return false;" class="indent more-label" id="log-more-label" >Show more... >></a>
    <a href="" onClick="showLess('log'); return false;" class="indent more-label" id="log-less-label" ><< Show less</a>
    <div class="indent more-content" id="log-more-content"><br />
<h2>Name</h2>
<p class="indent">log</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The "log" function computes the logarithm of a number, given a particular base.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             log( RealPos x, RealPos base )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The number whose logarithm needs to be computed.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >base</b>
<div class="indent">
The base for computing the logarithm.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # Let's compute the logarithm of a few numbers, changing the base
             log(1, base=exp(1))
             log(1, base=2)
             log(2, base=2)
             log(4, base=2)
             log(8, base=2)
             log(10, base=10)
             log(100, base=10)
             log(100, base=10.5)
             # Let's see what happens when we go beyond the definition range of log
             log(-1, base=10)
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ln'); return false;">ln</a><br />
<a href="" onClick="navigateHelp('exp'); return false;">exp</a><br />
<a href="" onClick="navigateHelp('power'); return false;">power</a><br />
</p><br />
</div>
</div>

<div id="ls">
    <span onClick="navigateHelp('ls');">
        <a href=""><h2>ls</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This function shows the content of the workspace.
            </p></div>    
    <a href="" onClick="showMore('ls'); return false;" class="indent more-label" id="ls-more-label" >Show more... >></a>
    <a href="" onClick="showLess('ls'); return false;" class="indent more-label" id="ls-less-label" ><< Show less</a>
    <div class="indent more-content" id="ls-more-content"><br />
<h2>Name</h2>
<p class="indent">ls</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This function shows the content of the workspace.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             ls( Bool all = FALSE )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >all</b>
<div class="indent">
Should all, even functions, be shown.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
FALSE<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            The list functions shows all the variables in the current workspace.
            You can also see all the functions available if you use ls(all=TRUE)
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                # now we have an empty workspace
                ls()
                # next wee add a variable
                a <- 1
                # and we can see it
                ls()
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('clear'); return false;">clear</a><br />
</p><br />
</div>
</div>

<div id="mapTree">
    <span onClick="navigateHelp('mapTree');">
        <a href=""><h2>mapTree</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Finds the tree with the highest posterior probability.
            </p></div>    
    <a href="" onClick="showMore('mapTree'); return false;" class="indent more-label" id="mapTree-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mapTree'); return false;" class="indent more-label" id="mapTree-less-label" ><< Show less</a>
    <div class="indent more-content" id="mapTree-more-content"><br />
<h2>Name</h2>
<p class="indent">mapTree</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Finds the tree with the highest posterior probability.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             mapTree( TreeTrace treetrace, String file )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >treetrace</b>
<div class="indent">
The tree trace in which we will search for the tree with the highest posterior probability.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
TreeTrace<br />
</div>
<br />
<b >file</b>
<div class="indent">
A description of the argument.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            This function searches through the tree trace and finds the topology with the most occurrences (highest posterior probability).
            Then, we use the mean branch lengths or node ages to build the tree.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
             treetrace = readTreeTrace("output/primates.trees", treetype="non-clock")
             mapTree(treetrace,"output/primates_map.tree")
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('readTreeTrace'); return false;">readTreeTrace</a><br />
</p><br />
</div>
</div>

<div id="mean">
    <span onClick="navigateHelp('mean');">
        <a href=""><h2>mean</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
                The 'mean' function computes the arithmetic mean of a vector of real numbers.
                The arithmetic mean simply computes the average by summing all the values together and divides them by the number of values.
            </p></div>    
    <a href="" onClick="showMore('mean'); return false;" class="indent more-label" id="mean-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mean'); return false;" class="indent more-label" id="mean-less-label" ><< Show less</a>
    <div class="indent more-content" id="mean-more-content"><br />
<h2>Name</h2>
<p class="indent">mean</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                The 'mean' function computes the arithmetic mean of a vector of real numbers.
                The arithmetic mean simply computes the average by summing all the values together and divides them by the number of values.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             mean( Real[] x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
This is the vector of real number.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real[]<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
             # compute the mean of some random draws from a normal distribution
             x <- rnorm(n=10, 5,1)
             mean(x)
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('sum'); return false;">sum</a><br />
<a href="" onClick="navigateHelp('var'); return false;">var</a><br />
</p><br />
</div>
</div>

<div id="normalize">
    <span onClick="navigateHelp('normalize');">
        <a href=""><h2>normalize</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'normalize' scales all vector elements by some constant such that their new average equals 'sum'.
            </p></div>    
    <a href="" onClick="showMore('normalize'); return false;" class="indent more-label" id="normalize-more-label" >Show more... >></a>
    <a href="" onClick="showLess('normalize'); return false;" class="indent more-label" id="normalize-less-label" ><< Show less</a>
    <div class="indent more-content" id="normalize-more-content"><br />
<h2>Name</h2>
<p class="indent">normalize</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'normalize' scales all vector elements by some constant such that their new average equals 'sum'.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             normalize( RealPos[] x, RealPos sum = 1 )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
A description of the argument.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos[]<br />
</div>
<br />
<b >sum</b>
<div class="indent">
A description of the argument.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            'normalize' computes the mean of 'x', divides all elements in 'x' by that mean, then multiplies by the value 'sum'. This results in the mean of 'normalize(x)' equaling 'sum'.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > x <- [1, 1, 3]
            > y := normalize(x, 4.0)
            > y
               [ 2, 4, 6 ]

            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
</div>

<div id="power">
    <span onClick="navigateHelp('power');">
        <a href=""><h2>power</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'power' raises 'base' to the 'exponent' power
            </p></div>    
    <a href="" onClick="showMore('power'); return false;" class="indent more-label" id="power-more-label" >Show more... >></a>
    <a href="" onClick="showLess('power'); return false;" class="indent more-label" id="power-less-label" ><< Show less</a>
    <div class="indent more-content" id="power-more-content"><br />
<h2>Name</h2>
<p class="indent">power</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'power' raises 'base' to the 'exponent' power
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             power( Real[] base, Real exponent )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >base</b>
<div class="indent">
The exponent's base value(s)<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real or Real[]<br />
</div>
<br />
<b >exponent</b>
<div class="indent">
The exponent's power<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                'power' returns the value(s) of 'base' raised to the power 'exponent'. If 'base' is a vector, then 'power' is applied element-wise to 'base'. When 'base' is RealPos, all values of 'exponent' are accepted. For negative Real values of 'base', values of 'exponent' that give non-complex powers are valid. In general, 'power(base,exponent)' is equivalent to applying the '^' operator, e.g. 'base^exponent'. 
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > power(0.01, 0.5)
                0.1
            > power([1,2,3], 2)
                [ 1, 4, 9 ]
            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('sqrt'); return false;">sqrt</a><br />
</p><br />
</div>
</div>

<div id="quit">
    <span onClick="navigateHelp('quit');">
        <a href=""><h2>quit</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Terminates the currently running instance of RevBayes.
            </p></div>    
    <a href="" onClick="showMore('quit'); return false;" class="indent more-label" id="quit-more-label" >Show more... >></a>
    <a href="" onClick="showLess('quit'); return false;" class="indent more-label" id="quit-less-label" ><< Show less</a>
    <div class="indent more-content" id="quit-more-content"><br />
<h2>Name</h2>
<p class="indent">quit</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Terminates the currently running instance of RevBayes.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             quit()
             </p>
<h2>Details</h2>
<p class="indent"><p class="indent">
                Quits RevBayes. All the current variables in the workspace will be lost.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                # if you really want to quit
                q()
                # you can always start again later ...
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
</div>
</div>

<div id="rcat">
    <span onClick="navigateHelp('rcat');">
        <a href=""><h2>rcat</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This function computes the probability mass of an observation x from a
            categorical distribution. The probability of the outcomes is specified
            by the simplex parameter p. The outcomes are labeled from 1 to K, and
            each outcome i has the probability given by element i in the simplex p.
            </p></div>    
    <a href="" onClick="showMore('rcat'); return false;" class="indent more-label" id="rcat-more-label" >Show more... >></a>
    <a href="" onClick="showLess('rcat'); return false;" class="indent more-label" id="rcat-less-label" ><< Show less</a>
    <div class="indent more-content" id="rcat-more-content"><br />
<h2>Name</h2>
<p class="indent">rcat</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This function computes the probability mass of an observation x from a
            categorical distribution. The probability of the outcomes is specified
            by the simplex parameter p. The outcomes are labeled from 1 to K, and
            each outcome i has the probability given by element i in the simplex p.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             dcat( Natural x, Simplex p, Bool log = true )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The observation.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
<b >p</b>
<div class="indent">
The probability of the outcomes.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
</div>
<br />
<b >log</b>
<div class="indent">
If TRUE, the probability p is given as log(p).<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            See dnCat for details.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
             # Compute the density of each of the outcomes
             # in a categorical distribution with the outcome
             # probabilities being 0.1, 0.2, 0.3, and 0.4,
             # respectively. Note that the outcomes are
             # in the set { 1, 2, 3, 4 }
             x <- v( 1, 2, 3, 4 )
             for ( i in 1:x.size() )
                y[i] <- dcat( x[i], simplex( 0.1, 0.2, 0.3, 0.4 ), FALSE )
             y
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Fredrik Ronquist</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnCat'); return false;">dnCat</a><br />
<a href="" onClick="navigateHelp('rcat'); return false;">rcat</a><br />
</p><br />
</div>
</div>

<div id="rep">
    <span onClick="navigateHelp('rep');">
        <a href=""><h2>rep</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
                'rep' creates a vector of 'n' copies of the value 'x'.
            </p></div>    
    <a href="" onClick="showMore('rep'); return false;" class="indent more-label" id="rep-more-label" >Show more... >></a>
    <a href="" onClick="showLess('rep'); return false;" class="indent more-label" id="rep-less-label" ><< Show less</a>
    <div class="indent more-content" id="rep-more-content"><br />
<h2>Name</h2>
<p class="indent">rep</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                'rep' creates a vector of 'n' copies of the value 'x'.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             rep( RealPos x, Natural n )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The value to be replicated<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >n</b>
<div class="indent">
The number of replicated values<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            'rep' creates a vector of 'n' elements, each with value 'x', preserving the type of 'x' in the returned vector.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > rep(0.1, 3)
               [ 0.1, 0.1, 0.1 ]
            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('seq'); return false;">seq</a><br />
<a href="" onClick="navigateHelp('simplex'); return false;">simplex</a><br />
<a href="" onClick="navigateHelp('v'); return false;">v</a><br />
</p><br />
</div>
</div>

<div id="round">
    <span onClick="navigateHelp('round');">
        <a href=""><h2>round</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'round' rounds the value 'x' to the nearest whole number.
            </p></div>    
    <a href="" onClick="showMore('round'); return false;" class="indent more-label" id="round-more-label" >Show more... >></a>
    <a href="" onClick="showLess('round'); return false;" class="indent more-label" id="round-less-label" ><< Show less</a>
    <div class="indent more-content" id="round-more-content"><br />
<h2>Name</h2>
<p class="indent">round</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'round' rounds the value 'x' to the nearest whole number.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             round( RealPos x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The value<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                'round' rounds the number 'x' to the nearest whole number. If 'x' is RealPos, 'round(x)' returns a Natural number. If 'x' is Real, 'round(x)' returns an Integer. 
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > round(1.499)
                1
            > round(1.5)
                2 
            > round(1.999)
                2
            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ceil'); return false;">ceil</a><br />
<a href="" onClick="navigateHelp('floor'); return false;">floor</a><br />
<a href="" onClick="navigateHelp('trunc'); return false;">trunc</a><br />
</p><br />
</div>
</div>

<div id="simplex">
    <span onClick="navigateHelp('simplex');">
        <a href=""><h2>simplex</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'simplex' constructs a Simplex object, which functions as a Vector of RealPos whose elements sum to one.
            </p></div>    
    <a href="" onClick="showMore('simplex'); return false;" class="indent more-label" id="simplex-more-label" >Show more... >></a>
    <a href="" onClick="showLess('simplex'); return false;" class="indent more-label" id="simplex-less-label" ><< Show less</a>
    <div class="indent more-content" id="simplex-more-content"><br />
<h2>Name</h2>
<p class="indent">simplex</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'simplex' constructs a Simplex object, which functions as a Vector of RealPos whose elements sum to one.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
            simplex( RealPos[] x )
            simplex( RealPos x1, RealPos x2, ... )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
List of values to construct a Simplex<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos[]<br />
</div>
<br />
<b >...</b>
<div class="indent">
Values to construct a Simplex<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos, RealPos, ...<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                'simplex' constructs a Simplex object from either a RealPos[] vector or from more than two RealPos values. Whenever a simplex value is assigned, each Simplex element is normalized by the sum of all Simplex elements, which guarantees the sum of all assigned Simplex values sums to one at all times.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > x <- simplex([2,2,4])
            > x
                [ 0.25, 0.25, 0.5 ]
            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('v'); return false;">v</a><br />
</p><br />
</div>
</div>

<div id="sqrt">
    <span onClick="navigateHelp('sqrt');">
        <a href=""><h2>sqrt</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'sqrt' computes the square root of a real positive number, 'x'.
            </p></div>    
    <a href="" onClick="showMore('sqrt'); return false;" class="indent more-label" id="sqrt-more-label" >Show more... >></a>
    <a href="" onClick="showLess('sqrt'); return false;" class="indent more-label" id="sqrt-less-label" ><< Show less</a>
    <div class="indent more-content" id="sqrt-more-content"><br />
<h2>Name</h2>
<p class="indent">sqrt</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'sqrt' computes the square root of a real positive number, 'x'.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             sqrt( RealPos x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The value<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            'sqrt' computes the square root of the real positive number, 'x', and returns the square-root as a real positive number. 'sqrt(x)' is equivalent to raising 'x' to the power 0.5.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > sqrt(2)
                1.41421
            > sqrt(2) == 2^0.5
                true
            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('power'); return false;">power</a><br />
</p><br />
</div>
</div>

<div id="trunc">
    <span onClick="navigateHelp('trunc');">
        <a href=""><h2>trunc</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'trunc' truncates 'x', dropping any numbers past the decimal point.
            </p></div>    
    <a href="" onClick="showMore('trunc'); return false;" class="indent more-label" id="trunc-more-label" >Show more... >></a>
    <a href="" onClick="showLess('trunc'); return false;" class="indent more-label" id="trunc-less-label" ><< Show less</a>
    <div class="indent more-content" id="trunc-more-content"><br />
<h2>Name</h2>
<p class="indent">trunc</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'trunc' truncates 'x', dropping any numbers past the decimal point.
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
             trunc( RealPos x )
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The value<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            'trunc' truncates the number 'x' by dropping all digits to the right of the decimal point, rounding 'x' towards zero. For positive values, 'trunc' is equivalent to 'floor'. For negative values, 'trunc' is equivalent to 'ceil'.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > trunc(1.2)
                1.0
            > trunc(-1.2)
                -1.0
            </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('ceil'); return false;">ceil</a><br />
<a href="" onClick="navigateHelp('floor'); return false;">floor</a><br />
<a href="" onClick="navigateHelp('round'); return false;">round</a><br />
</p><br />
</div>
</div>

<div id="v">
    <span onClick="navigateHelp('v');">
        <a href=""><h2>v</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="function" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            'v' creates a vector of the elements '...'
            </p></div>    
    <a href="" onClick="showMore('v'); return false;" class="indent more-label" id="v-more-label" >Show more... >></a>
    <a href="" onClick="showLess('v'); return false;" class="indent more-label" id="v-less-label" ><< Show less</a>
    <div class="indent more-content" id="v-more-content"><br />
<h2>Name</h2>
<p class="indent">v</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            'v' creates a vector of the elements '...'
            </p>
</p>
<h2>Usage</h2>
<p class="indent">
            v( Real, Real, ... )
            v( Real[], Real[], ... )
            ...
             </p>
<h2 >Arguments</h2>
<div class="indent">
<b ></b>
<div class="indent">
A description of the argument.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
Any<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                'v' creates a vector of the elements '...', which are objects of a common base type. Vector elements may themselves be vectors.
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
            > # create a vector, Natural[]
            > x <- v(1,2,3)
            > x <- x + 1
            > x
               [ 2, 3, 4 ]
            > y <- v(2,4,6)
            > # create a vector of Natural[][]
            > z <- v(x,y)
            > z
               [ [ 2, 3, 4 ], [ 2, 4, 6 ] ]
            > z[0]
               [ 2, 3, 4 ]
             </pre>
</p>
<h2>Author</h2>
<p class="indent">Michael Landis</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('simplex'); return false;">simplex</a><br />
</p><br />
</div>
</div>

</div>
            
            <br/>
            <br/>
            <div class="section-label"><strong><em>Distribution</em></strong></div>            
            <div id="distributionEntries"><div id="dnACLN">
    <span onClick="navigateHelp('dnACLN');">
        <a href=""><h2>dnACLN</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is the constructor function for the autocorrelated, log-normally distributed node rates
            on a tree. This model was described in Thorne et al. (1998), Kishino et al. (2001), and Thorne &
            Kishino (2002).
            </p></div>    
    <a href="" onClick="showMore('dnACLN'); return false;" class="indent more-label" id="dnACLN-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnACLN'); return false;" class="indent more-label" id="dnACLN-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnACLN-more-content"><br />
<h2>Name</h2>
<p class="indent">dnACLN</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is the constructor function for the autocorrelated, log-normally distributed node rates
            on a tree. This model was described in Thorne et al. (1998), Kishino et al. (2001), and Thorne &
            Kishino (2002).
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnACLN( TimeTree tree, RealPos sigma, Real rootRate, RealPos scaleValue )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >tree</b>
<div class="indent">
The time-tree on which the node rates are mapped.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
TimeTree<br />
</div>
<br />
<b >sigma</b>
<div class="indent">
The variance parameter controlling the degree of autocorrelation among node rates.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >rootRate</b>
<div class="indent">
The rate value at the root.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
<b >scaleValue</b>
<div class="indent">
The base-rate of the process.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This distribution gives a vector of RealPos values that are mapped to the non-root
            	nodes of a tree. A separate function (aveRateOnBranch) taking this vector and the other ACLN parameters
            	is used to compute the branch-rates. 
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
	## read in the tree 
	T <- readTrees("data/bears_dosReis.tre")[1]

	### Load the sequence alignment
	D <- readDiscreteCharacterData(file="data/bears_irbp.nex")
	
	### get helpful variables from the data
	n_sites <- D.nchar(1)
	n_taxa <- D.ntaxa()
	n_branches <- 2 * n_taxa - 2
	names <- D.names()

	### initialize an iterator for the moves vector
	mi = 1

	### Specify Yule (pure birth) tree prior
	birth_rate <- 0.1
	root_time ~ dnExponential(0.1, offset=35.0) 
	timetree ~ dnBDP(lambda=birth_rate, mu=0.0, rho=1.0, rootAge=root_time, samplingStrategy="uniform", condition="nTaxa", nTaxa=n_taxa, names=names)
	
	### set up the base-rate parameter of the ACLN model 
	base_rate <- 0.01

	### set up the variance parameter of the ACLN model 
	nu <- 0.1
	
	### the clock rate at the root of the tree
	root_rate ~ dnGamma(2.0, 4.0)

	### a vector of the rates at each node
	### from the ACLN model, which is conditioned on the tree, nu, the root rate, and base rate
	node_rates ~ dnACLN(timetree, nu, root_rate, base_rate)
	
	### the rate at each branch is the average of the rate at each node separated by that branch
	### a special core function is used to calculate the rate for branch i
	for(i in 1:n_branches){
		branch_rates[i] := aveRateOnBranch(node_rates, timetree, root_rate, base_rate, index=i)
	}
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    Kishino H, Thorne JL, Bruno W. 2001. Performance of a divergence time estimation method under a probabilistic model of rate evolution. Mol Biol Evol. 18:352–361.
                <br />
<br />
<br />

                    Rannala B, Yang Z. 2003. Bayes estimation of species divergence times and ancestral population sizes using DNA sequences from multiple loci. Genetics 164:1645–1656. 
                <br />
<br />
<br />

                    Thorne J, Kishino H. 2002. Divergence time and evolutionary rate estimation with multilocus data. Syst Biol. 51:689–702.
                <br />
<br />
<br />

                    Thorne J, Kishino H, Painter IS. 1998. Estimating the rate of evolution of the rate of molecular evolution. Mol Biol Evol. 15:1647–1657.
                <br />
<br />
<br />

                    Yang Z, Rannala B. 2006. Bayesian estimation of species divergence times under a molecular clock using multiple fossil calibrations with soft bounds. Mol Biol Evol. 23:212–226.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Tracy A. Heath</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('aveRateOnBranch'); return false;">aveRateOnBranch</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnBDP">
    <span onClick="navigateHelp('dnBDP');">
        <a href=""><h2>dnBDP</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
                The constant-rate birth-death process specifies a probability distribution on time trees. The birth (speciation) and death (extinction)
                rates are assumed to be constant over time. Unlike some implementations, cBDP does not constrain the birth rate to be larger than the death
                rate.
            </p></div>    
    <a href="" onClick="showMore('dnBDP'); return false;" class="indent more-label" id="dnBDP-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnBDP'); return false;" class="indent more-label" id="dnBDP-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnBDP-more-content"><br />
<h2>Name</h2>
<p class="indent">dnBDP</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                The constant-rate birth-death process specifies a probability distribution on time trees. The birth (speciation) and death (extinction)
                rates are assumed to be constant over time. Unlike some implementations, cBDP does not constrain the birth rate to be larger than the death
                rate.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnBDP( RealPos lambda, RealPos mu = 0, RealPos origin = NULL, RealPos rootAge = NULL, Probability rho = 1, String samplingStrategy = "uniform", String condition = "survival", Natural nTaxa, String[] names, Clade[] constraints = [  ] )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >lambda</b>
<div class="indent">
Birth (speciation) rate.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >mu</b>
<div class="indent">
Death (extinction) rate.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
0<br />
</div>
<br />
<b >origin</b>
<div class="indent">
The time of the process since the origin.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
NULL<br />
</div>
<br />
<b >rootAge</b>
<div class="indent">
The time of the process since the root (most recent common ancestor).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
NULL<br />
</div>
<br />
<b >rho</b>
<div class="indent">
The taxon sampling probability.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Probability<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >samplingStrategy</b>
<div class="indent">
The sampling strategy used if the taxon sampling is incomplete.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
<span class="argument">Default value</span>
"uniform"<br />
</div>
<br />
<b >condition</b>
<div class="indent">
The condition of the process. For details, see below<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
<span class="argument">Default value</span>
"survival"<br />
</div>
<br />
<b >nTaxa</b>
<div class="indent">
The number of taxa. This is only needed for simulation/initialization purposes when a tree with exactly n taxa is needed in the model.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
<b >names</b>
<div class="indent">
The taxon labels that will be applied to the simulated tree.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String[]<br />
</div>
<br />
<b >constraints</b>
<div class="indent">
The topological constraints applied to the tree, i.e., conditioning on monphyly of a given clade.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Clade[]<br />
<span class="argument">Default value</span>
[  ]<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                    The implementation of the constant-rate birth-death process is based on Nee et al. (1994). The unconditioned process results in an improper probability
                    distribution, so the process needs to be conditioned on something. Possibilities include to condition the process on the time (duration), on survival
                    (non-extinction), on obtaining a tree, or on obtaining the number of observed taxa. The condition of the process is controlled by the 'condition'
                    parameter.
                </p>
<p class="indent">
                    The process gives rise to trees with extinct branches, but observed trees typically do not include extinct branches. As implemented here, we
                    assume that it is trees without extinct branches, so-called reconstructed trees, that we observe. The probability distribution is modified accordingly.
                </p>
<p class="indent">
                    It is commonly the case that the observed tree does not include all tips. The included tips may have been sampled uniformly at random from the available
                    complete tree. A particularly common bias is that the investigator has tried to maximize the diversity in the sampled tree, by including tips that are as
                    distantly related to each other as possible. This is referred to as diversified sampling. It is critical to model the sampling process
                    correctly to obtain accurate estimates of the birth and death rates (Hoehna et al. 2011); particularly the estimate of the death rate is sensitive to
                    violations of the sampling assumptions.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    trueTree <- readTrees("data/primates.tree")[1]
                    
                    names <- trueTree.names()
                    taxaCount <- names.size()
                    
                    # Birth-Death process priors
                    speciation ~ dnExponential(0.1)
                    extinction ~ dnExponential(0.1)
                    org ~ dnUniform(0.0,50.0)
                    
                    # moves for the birth-death parameters
                    moves[1] = mvScale(speciation, lambda=1, tune=true, weight=2)
                    moves[2] = mvScale(extinction, lambda=1, tune=true, weight=2)
                    moves[3] = mvSlide(org, delta=1.0, weight=2.0)
                    
                    # tree from birth-death process
                    tau ~ dnBDP(origin=org, lambda=speciation, mu=extinction, nTaxa=taxaCount, names=names)
                    
                    tau.clamp( trueTree )
                    
                    monitors[1] = modelmonitor(filename="testBirthDeath.log",printgen=10, separator = "	")
                    monitors[2] = screenmonitor(printgen=1000, separator = "	", speciation)
                    
                    mymodel = model(tau)
                    
                    mymcmc = mcmc(mymodel, monitors, moves)
                    
                    mymcmc.burnin(generations=20000,tuningInterval=100)
                    mymcmc.run(generations=200000)
                    mymcmc.operatorSummary()
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    Nee, S., May, R.M., and Harvey, P.H. 1994. The Reconstructed Evolutionary Process. Philosophical Transactions: Biological Sciences, 344, 305-311
                <br />
http://rstb.royalsocietypublishing.org/content/344/1309/305.short<br />
10.1098/rstb.1994.0068<br />

                    Höhna, S., Stadler, T., Ronquist, F, and Britton, T. 2011. Inferring speciation and extinction rates under different species sampling schemes. Molecular Biology and Evolution, 28, 2577-2589.
                <br />
http://mbe.oxfordjournals.org/content/28/9/2577.short<br />
10.1093/molbev/msr095<br />
</p><br />
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnBDPSerial'); return false;">dnBDPSerial</a><br />
<a href="" onClick="navigateHelp('dnBDPSkySerial'); return false;">dnBDPSkySerial</a><br />
<a href="" onClick="navigateHelp('dnYuleDivDep'); return false;">dnYuleDivDep</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnBernoulli">
    <span onClick="navigateHelp('dnBernoulli');">
        <a href=""><h2>dnBernoulli</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            A Bernoulli-distributed random variable takes the value 1 with probability
            p and the value 0 with probability 1-p.
            </p></div>    
    <a href="" onClick="showMore('dnBernoulli'); return false;" class="indent more-label" id="dnBernoulli-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnBernoulli'); return false;" class="indent more-label" id="dnBernoulli-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnBernoulli-more-content"><br />
<h2>Name</h2>
<p class="indent">dnBernoulli</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            A Bernoulli-distributed random variable takes the value 1 with probability
            p and the value 0 with probability 1-p.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                    dnBernoulli( Probability prob )
                    dbernoulli( Natural x, Probability prob )
                    pbernoulli( Natural q, Probability prob )
                    qbernoulli( Probability p, Probability prob )
                    rbernoulli( Natural n, Probability prob )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >prob</b>
<div class="indent">
The probability of success on the single trial.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Probability<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This function was among the easiest in RevBayes to implement. 
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    p ~ dnBeta(1.0,1.0)
                    x ~ dnBernoulli(p)
                    x.clamp(1)
                    moves[1] = mvSlide(p, delta=0.1, weight=1.0)
                    monitors[1] = screenmonitor(printgen=1000, separator = "	", speciation)
                    mymodel = model(p)
                    mymcmc = mcmc(mymodel, monitors, moves)
                    mymcmc.burnin(generations=20000,tuningInterval=100)
                    mymcmc.run(generations=200000)
                    
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    The Bernoulli distribution is attributed to the Swiss scientist Jacob Bernoulli.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">John Huelsenbeck</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnBinomial'); return false;">dnBinomial</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnBeta">
    <span onClick="navigateHelp('dnBeta');">
        <a href=""><h2>dnBeta</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The Beta probability distribution.
            </p></div>    
    <a href="" onClick="showMore('dnBeta'); return false;" class="indent more-label" id="dnBeta-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnBeta'); return false;" class="indent more-label" id="dnBeta-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnBeta-more-content"><br />
<h2>Name</h2>
<p class="indent">dnBeta</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The Beta probability distribution.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnBeta( RealPos alpha, RealPos beta )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >alpha</b>
<div class="indent">
The alpha parameter.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >beta</b>
<div class="indent">
The beta parameter.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The Beta distribution is a probability distribution defined on 0 to 1.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # we create a variable x from a beta distribution
                 x ~ dnBeta(1,1)
                 # this variable could be a parameter from a binomial distribution
                 y ~ dnBinomial(n=10,p)
                 # now you could clamp the variable y and run an MCMC on it ...
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dBeta'); return false;">dBeta</a><br />
<a href="" onClick="navigateHelp('rBeta'); return false;">rBeta</a><br />
<a href="" onClick="navigateHelp('dnDirichlet'); return false;">dnDirichlet</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnBinomial">
    <span onClick="navigateHelp('dnBinomial');">
        <a href=""><h2>dnBinomial</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Binomial probability distribution of x successes in n trials.
            </p></div>    
    <a href="" onClick="showMore('dnBinomial'); return false;" class="indent more-label" id="dnBinomial-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnBinomial'); return false;" class="indent more-label" id="dnBinomial-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnBinomial-more-content"><br />
<h2>Name</h2>
<p class="indent">dnBinomial</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Binomial probability distribution of x successes in n trials.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnBinomial( Probability p, Natural n )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >p</b>
<div class="indent">
The probability of success.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Probability<br />
</div>
<br />
<b >n</b>
<div class="indent">
The number of trials.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The binomial probability distribution defines the number of success in n trials,
                where each trial has the same success probability p. The probability is given by
                (n choose x) p^(x) * (1-p)^(n-p)
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # we create a variable x from a beta distribution
                    x ~ dnBeta(1,1)
                    # this variable could be a parameter from a binomial distribution
                    y ~ dnBinomial(n=10,p)
                    # now you could clamp the variable y and run an MCMC on it ...
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnBeta'); return false;">dnBeta</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnBrownian">
    <span onClick="navigateHelp('dnBrownian');">
        <a href=""><h2>dnBrownian</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is the constructor function for the univariate Brownian model of quantitative trait evolution along a time-calibrated phylogeny (Felsenstein, 1973). 
            </p></div>    
    <a href="" onClick="showMore('dnBrownian'); return false;" class="indent more-label" id="dnBrownian-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnBrownian'); return false;" class="indent more-label" id="dnBrownian-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnBrownian-more-content"><br />
<h2>Name</h2>
<p class="indent">dnBrownian</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is the constructor function for the univariate Brownian model of quantitative trait evolution along a time-calibrated phylogeny (Felsenstein, 1973). 
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnBrownian( TimeTree tree, RealPos sigma, Real drift = 0 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >tree</b>
<div class="indent">
A time-calibrated phylogeny.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
TimeTree<br />
</div>
<br />
<b >sigma</b>
<div class="indent">
The evolutionary rate parameter (variance per unit of time).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >drift</b>
<div class="indent">
The drift (systematic trend) of the process.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
<span class="argument">Default value</span>
0<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                A Brownian motion is entirely determined by its initial value X(0), its evolutionary rate of change sigma (a positive real number) and its drift d (a real number). Along a given time interval (0,T), the value of X at time T is normally distributed of mean X(0)+ d*T and variance sigma^2*T. In this implementation, the value at the root, X(0), has an improper uniform distribution.
                </p>
<p class="indent">
                The process can be clamped at the tips of the phylogeny to the values specified by a continuous data matrix. Values of X at the interior nodes of the phylogeny are usually integrated out (analytically). Here, in contrast, ancestral values are explicitly sampled by MCMC. This can be useful in the context of molecular comparative analyses, although this results in less efficient MCMC mixing.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # load a tree
                 trueTree = readTrees("data/chronoplac40.tree")[1]

                 # put a log-uniform distribution on the evolutionary rate parameter 
                 sigma ~ dnLogUniform(min=0.001, max=1000)

                 # create a Brownian process without drift
                 X ~ dnBrownian(tree=trueTree, sigma=sigma) 

                 # load quantitative trait data
                 contData = readContinuousCharacterData("data/plac40lhtlog.nex")

                 # clamp Brownian process to second column of quantitative trait data (log body mass in that case)
                 X.clampAt(contData,2)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    Felsenstein J, 1973. Maximum likelihood estimation of evolutionary trees from continuous characters. Am. J Hum. Genet. 25:471-492.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnBrownianMultiVariate">
    <span onClick="navigateHelp('dnBrownianMultiVariate');">
        <a href=""><h2>dnBrownianMultiVariate</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is the constructor function for the multivariate Brownian model of quantitative trait evolution along a time-calibrated phylogeny (Felsenstein, 1985). 
            </p></div>    
    <a href="" onClick="showMore('dnBrownianMultiVariate'); return false;" class="indent more-label" id="dnBrownianMultiVariate-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnBrownianMultiVariate'); return false;" class="indent more-label" id="dnBrownianMultiVariate-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnBrownianMultiVariate-more-content"><br />
<h2>Name</h2>
<p class="indent">dnBrownianMultiVariate</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is the constructor function for the multivariate Brownian model of quantitative trait evolution along a time-calibrated phylogeny (Felsenstein, 1985). 
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnBrownianMultiVariate( TimeTree tree, RealSymmetricMatrix sigma )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >tree</b>
<div class="indent">
A time-calibrated phylogeny.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
TimeTree<br />
</div>
<br />
<b >sigma</b>
<div class="indent">
The covariance per unit of time.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealSymmetricMatrix<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                A multivariate Brownian motion of dimension D is parameterized by its initial value X(0) (a vector of dimension D) and its rate of change Sigma(a DxD symmmetric definite positive covariance matrix). Along a given time interval (0,T), the value of X at time T is multivariate normal of mean X(0) and covariance sigma*T. The value at the root, X(0), has an improper uniform distribution.
                </p>
<p class="indent">
                The process can be clamped at the tips of the phylogeny to the values specified by a continuous data matrix. This should be done trait by trait(see below). Values of X at the interior nodes of the phylogeny are usually integrated out (analytically). Here, in contrast, ancestral values are explicitly sampled by MCMC. This can be useful in the context of molecular comparative analyses, although this results in less efficient MCMC mixing.
                </p>
<p class="indent">
                A mathematically convenient prior for the covariance matrix of a multivariate Brownian motion is the inverse-Wishart distribution (see dnInvWishart and example below). 
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # load a tree
                 trueTree = readTrees("data/chronoplac40.tree")[1]

                 # load quantitative trait data
                 contData = readContinuousCharacterData("data/plac40lhtlog.nex")
                 # get number of traits (dimension of multivariate Brownian process)
                 nTraits <- contData.nchar()[1]

                 # use an inverse-Wishart prior distribution for the rate of the Brownian process
                 # of parameters kappa * Identity and df = nTraits+2 degrees of freedom
                 # where kappa is itself a scaling factor
                 kappa ~ dnLogUniform(min=0.001,max=1000)
                 # can be useful to start from a reasonable value for kappa
                 kappa.setValue(0.1)
                 Sigma ~ dnInvWishart(dim=nTraits, kappa=kappa, df=nTraits+2)

                 # multivariate Brownian process along the tree
                 X ~ dnBrownianMultiVariate(trueTree,sigma=Sigma)

                 # clamp the process at the tips of the tree, at the trait values observed in extant taxa
                 # this has to be done trait by trait
                 for (i in 1:nTraits)	{
                       # the ith entry of the Brownian process should map to the ith. quantitative trait
                       X.clampAt(contData,i,i)
                 }
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    Felsenstein F, 1985. Phylogenies and the comparative method. Am. Nat. 125:1-15.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnCat">
    <span onClick="navigateHelp('dnCat');">
        <a href=""><h2>dnCat</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The categorical distribution, sometimes referred to as the generalized Bernoulli
            distribution. It describes the probability of one of K different outcomes,
            labeled from 1 to K, with each outcome probability separately specified.
            </p></div>    
    <a href="" onClick="showMore('dnCat'); return false;" class="indent more-label" id="dnCat-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnCat'); return false;" class="indent more-label" id="dnCat-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnCat-more-content"><br />
<h2>Name</h2>
<p class="indent">dnCat</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The categorical distribution, sometimes referred to as the generalized Bernoulli
            distribution. It describes the probability of one of K different outcomes,
            labeled from 1 to K, with each outcome probability separately specified.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnCat( Simplex p )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >p</b>
<div class="indent">
Vector of outcome probabilities<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                    The argument to the constructor is a simplex containing the
                    probabilities of the outcomes. The outcomes are labeled from
                    1 to K, where K is the number of elements in the simplex.
                    Outcome i has probability specified by component i
                    in the simplex.
                </p>
<p class="indent">
                    A typical scenario where a categorical variable is used
                    is in the definition of a variable drawn from a mixture.
                    A vector of mixture components is set up first, and then
                    a stochastic variable drawn from a categorical distribution
                    is used as an index in a deterministic assignment that
                    points to a component in the mixture. See example below.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # Define a stochastic variable x that is drawn from
                 # a categorical distribution with 4 categories, each
                 # category having the same probability, then examine
                 # the value of x.
                 x ~ dnCat( simplex(1,1,1,1) )
                 x

                 # Draw 10 values from the distribution and place them
                 # in a vector a, then examine a.
                 for ( i in 1:10 ) {
                    a[i] <- x
                    x.redraw()
                 }
                 a
                 
                 # Use x in defining a deterministic variable y taking
                 # on values from a mixture of RealPos values representing
                 # rates from a discretized scaled gamma distribution
                 # with four categories.
                 shape ~ dnExp( 10.0 )
                 rates := fnDiscretizeGamma( shape, shape, 4 )
                 y := rates[x]
                </pre>
</p>
<h2>Author</h2>
<p class="indent">Fredrik Ronquist</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dcat'); return false;">dcat</a><br />
<a href="" onClick="navigateHelp('rcat'); return false;">rcat</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnCoalMultiSpeciesConst">
    <span onClick="navigateHelp('dnCoalMultiSpeciesConst');">
        <a href=""><h2>dnCoalMultiSpeciesConst</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
                The 'dnCoalMultiSpeciesConst' distribution specifies a probability distribution on gene trees given a species time tree and a vector of effective population sizes, one per branch of the species tree.
                If only one effective population size is provided, all branches of the species tree are associated to the same effective population size.    
                In all cases, effective population sizes are assumed to be piecewise constant, i.e. along a given branch of the species tree, the effective population size does not vary. 
                A vector of taxa specifies how many leaves are expected in the gene trees, provides their names, and the species they correspond to. There can be several alleles per species. The number of alleles can vary across species.
            </p></div>    
    <a href="" onClick="showMore('dnCoalMultiSpeciesConst'); return false;" class="indent more-label" id="dnCoalMultiSpeciesConst-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnCoalMultiSpeciesConst'); return false;" class="indent more-label" id="dnCoalMultiSpeciesConst-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnCoalMultiSpeciesConst-more-content"><br />
<h2>Name</h2>
<p class="indent">dnCoalMultiSpeciesConst</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                The 'dnCoalMultiSpeciesConst' distribution specifies a probability distribution on gene trees given a species time tree and a vector of effective population sizes, one per branch of the species tree.
                If only one effective population size is provided, all branches of the species tree are associated to the same effective population size.    
                In all cases, effective population sizes are assumed to be piecewise constant, i.e. along a given branch of the species tree, the effective population size does not vary. 
                A vector of taxa specifies how many leaves are expected in the gene trees, provides their names, and the species they correspond to. There can be several alleles per species. The number of alleles can vary across species.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnCoalMultiSpeciesConst( TimeTree speciesTree, RealPos Ne, Taxon[] taxa )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >speciesTree</b>
<div class="indent">
The species tree, with species names as tips.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
TimeTree<br />
</div>
<br />
<b >Ne</b>
<div class="indent">
The effective population size or a vector of effective population sizes.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos, RealPos[]<br />
</div>
<br />
<b >taxa</b>
<div class="indent">
A vector of taxa corresponding to genes coming from the genomes of the species found in the species tree.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Taxon[]<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This distribution can be used for both inference and simulation.
                It was first proposed by Rannala and Yang (2003) and is widely used for species tree/gene tree inference.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # In this example, we first simulate gene trees given a species tree, and then we compute the probability of the gene trees given the species tree.
				# Number of species
				n_species = 4
				# Number of loci (in this example we just use 1)
				n_loci = 1 
				# Number of alleles per species
				n_alleles = v(1,2,1,1)

				# Species names
				for (i in 1:n_species) {
        			s_names[i] = "Species_"+i
				}


				# Species-Tree model
				# We simulate a species tree topology according to a birth-death process with arbitrary parameter values:
				speciation ~ dnExponential(10.0)
				extinction ~ dnExponential(10.0)
				tree_height ~ dnUnif(0,1.0)
				speciesTree ~ dnBDP(lambda=speciation, mu=extinction, origin=tree_height, nTaxa=n_species, names=s_names)

				# Gene-Tree multispecies coalescent model
				# Then we can use the multispecies coalescent model to generate gene trees. We simulate a set of gene trees, using one 
				# effective population size per branch, and after having constructed a map between species names and sequence names:

				# Build the mapping between sequence names and species names.
				totalNumberOfAlleles = 1
				for (i in 1:n_species) {
        			for (j in 1:n_alleles[i]) {
                		taxa[totalNumberOfAlleles] = taxon(taxonName=s_names[i]+"_"+j, speciesName=s_names[i])
                		totalNumberOfAlleles = totalNumberOfAlleles+1
        			}
				}

				# Set the effective population sizes
				n_branches = speciesTree.nnodes()
				for (i in 1:n_branches) {
					Ne[i] ~ dnGamma(shape=1.0,rate=1.0)
				}

				# Simulate gene trees
				for (i in 1:n_loci) {
        				# The gene tree from the multispecies coalescent process
        				locusTrees[i] ~ dnCoalMultiSpeciesConst(speciesTree=speciesTree, Ne=Ne, taxa=taxa)
				}
				
				# Now, we get the log probability of the gene tree given its parameters:
				print(locusTrees[1].lnProbability() )
				
				# And we can simulate another gene tree for this locus:
				locusTrees[1].redraw()
				
				# And recompute its log probability:
				print(locusTrees[1].lnProbability() )
				
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    Rannala B and Yang Z. 2003. "Bayes estimation of species divergence times and ancestral population sizes using DNA sequences from multiple loci.". Genetics, 164:4, 1645-1656.
                <br />
http://www.ncbi.nlm.nih.gov.gate1.inist.fr/pmc/articles/PMC1462670/<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('maximumTree'); return false;">maximumTree</a><br />
<a href="" onClick="navigateHelp('fnPomo'); return false;">fnPomo</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnDPP">
    <span onClick="navigateHelp('dnDPP');">
        <a href=""><h2>dnDPP</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This function returns a Dirichlet process prior (DPP) distribution. 
            </p></div>    
    <a href="" onClick="showMore('dnDPP'); return false;" class="indent more-label" id="dnDPP-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnDPP'); return false;" class="indent more-label" id="dnDPP-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnDPP-more-content"><br />
<h2>Name</h2>
<p class="indent">dnDPP</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This function returns a Dirichlet process prior (DPP) distribution. 
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnDPP( RealPos concentration, Distribution__<type> baseDistribution, Natural numElements )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >concentration</b>
<div class="indent">
The concentration parameter of the DPP (can be computed using fnDPPConcFromMean function).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >baseDistribution</b>
<div class="indent">
The DPP base distribution. This also gives the Rev type of the stochastic node created with this distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Distribution__Any (See details below)<br />
</div>
<br />
<b >numElements</b>
<div class="indent">
The number of elements distributed by the DPP.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This function constructs a Dirichlet process prior distribution. The DPP is used to model
                mixtures of data. For example, this distribution is used as a prior on lineage-specific
                substitution rates as a model for relaxing the assumption of a strict molecular clock (Heath et al. 2012).
                </p>
<p class="indent">
                In RevBayes this distribution is templated, meaning that it can be used for different types of
                parameters. Thus, depending on the type of the base distribution, the DPP can be a prior on elements 
                with Real, RealPos, Integer, Natural, Probability, or Simplex parameters. 
                The distribution constructor creates a stochastic node that is a vector of N elements of the type listed above,
                were N is the number of elements distributed under the DPP. 
                </p>
<p class="indent">
                The DPP is an infinite mixture model and takes 2 parameters: the concentration parameter (a) and 
                the base distribution (G_0). The concentration parameter influences the number of parameter categories
                and the base distribution is the prior distribution from which parameter values for each
                category are drawn. For more information see: Ferguson (1973); Antoniak (1974); Lartillot and Philippe (2004);
                Huelsenbeck et al. (2006); Ané et al. (2007); Huelsenbeck and Suchard (2007); Heath et al. (2012); 
                Heath (2012)
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
	 ## number of DPP elements
	 N <- 10
 
	 ## prior mean number of categories
	 k <- 8
 
	 ## concentration parameter of the DPP
	 a <- fnDPPConcFromMean(k, N)
 
	 ## G_0 distribution
	 G_0 = dnExponential(1.0)
 
	 ## DPP distribution constructior
	 rates ~ dnDPP(a, G_0, N)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
					Ané C, Larget B, Baum DA, Smith SD, Rokas A. 2007. Bayesian estimation of concordance among gene trees. Mol Biol Evol. 24: 412–426.
            	<br />
<br />
<br />

					Antoniak CE. 1974. Mixtures of Dirichlet processes with applications to non-parametric problems. Ann Stat. 2:1152–1174.
            	<br />
<br />
<br />

					Ferguson TS. 1973. A Bayesian analysis of some nonparametric problems. Ann Stat. 1:209–230.
            	<br />
<br />
<br />

					Heath TA, Holder MT, Huelsenbeck JP. 2012. A Dirichlet process prior for estimating lineage-specific substitution rates. Mol. Biol. Evol. 29(3):939–955. 
            	<br />
<br />
<br />

					Heath TA. 2012. A hierarchical Bayesian model for calibrating estimates of species divergence times. Syst. Biol. 61:793-809.
            	<br />
<br />
<br />

					Huelsenbeck JP, Jain S, Frost SWD, Pond SLK. 2006. A Dirichlet process model for detecting positive selection in protein-coding DNA sequences. Proc Natl Acad Sci U S A. 103:6263–6268.
            	<br />
<br />
<br />

					Huelsenbeck JP, Suchard M. 2007. A nonparametric method for accommodating and testing across-site rate variation. Syst Biol. 56:975–987.
            	<br />
<br />
<br />

					Lartillot N, Philippe H. 2004. A Bayesian mixture model for across-site heterogeneities in the amino-acid replacement process. Mol Biol Evol.21:1095–1109.
				<br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Tracy A. Heath</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnMixture'); return false;">dnMixture</a><br />
<a href="" onClick="navigateHelp('fnDPPConcFromMean'); return false;">fnDPPConcFromMean</a><br />
<a href="" onClick="navigateHelp('mvDPPAllocateAuxGibbs'); return false;">mvDPPAllocateAuxGibbs</a><br />
<a href="" onClick="navigateHelp('mvDPPGibbsConcentration'); return false;">mvDPPGibbsConcentration</a><br />
<a href="" onClick="navigateHelp('mvDPPScaleCatAllocateAux'); return false;">mvDPPScaleCatAllocateAux</a><br />
<a href="" onClick="navigateHelp('mvDPPScaleCatVals'); return false;">mvDPPScaleCatVals</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnDirichlet">
    <span onClick="navigateHelp('dnDirichlet');">
        <a href=""><h2>dnDirichlet</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Dirichlet probability distribution on a simplex.
            </p></div>    
    <a href="" onClick="showMore('dnDirichlet'); return false;" class="indent more-label" id="dnDirichlet-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnDirichlet'); return false;" class="indent more-label" id="dnDirichlet-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnDirichlet-more-content"><br />
<h2>Name</h2>
<p class="indent">dnDirichlet</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Dirichlet probability distribution on a simplex.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnDirichlet( RealPos[] alpha )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >alpha</b>
<div class="indent">
The weights of each category.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos[]<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The Dirichlet probability distribution is the generalization of the Beta distribution.
                A random variable from a Dirichlet distribution is a simplex, i.e., a vector of probabilities that sum to 1.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # lets get a draw from a Dirichlet distribution
                    a <- [1,1,1,1]   # we could also use rep(1,4)
                    b ~ dnDirichlet(a)
                    b
                    # let check if b really sums to 1
                    sum(b)
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnBeta'); return false;">dnBeta</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnExponential">
    <span onClick="navigateHelp('dnExponential');">
        <a href=""><h2>dnExponential</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Random variable that has an exponential distributed waiting time.
            </p></div>    
    <a href="" onClick="showMore('dnExponential'); return false;" class="indent more-label" id="dnExponential-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnExponential'); return false;" class="indent more-label" id="dnExponential-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnExponential-more-content"><br />
<h2>Name</h2>
<p class="indent">dnExponential</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Random variable that has an exponential distributed waiting time.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnExponential( RealPos lambda, Real offset )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >lambda</b>
<div class="indent">
The rate parameter (rate = 1/mean).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >offset</b>
<div class="indent">
This is an optional argument if you want to use the offset exponential distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This is one of the most basic distribution with probability density
                f(x) = lambda * exp(-lambda*x)
                and distribution function
                F(x) = 1 - exp(-lambda*x)
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # lets get some random draws
                    x <- rexp(1000,lambda=0.1)
                    # since the rate is 0.1 we expect the mean to be 10
                    mean(x)
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnGamma'); return false;">dnGamma</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnGamma">
    <span onClick="navigateHelp('dnGamma');">
        <a href=""><h2>dnGamma</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Gamma probability distribution for positive real numbers.
            </p></div>    
    <a href="" onClick="showMore('dnGamma'); return false;" class="indent more-label" id="dnGamma-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnGamma'); return false;" class="indent more-label" id="dnGamma-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnGamma-more-content"><br />
<h2>Name</h2>
<p class="indent">dnGamma</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Gamma probability distribution for positive real numbers.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnGamma( RealPos shape, RealPos rate )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >shape</b>
<div class="indent">
The shape parameter.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >rate</b>
<div class="indent">
The rate parameter.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The gamma distribution is the probability of the sum of exponentially distributed variables.
                Thus, it provides a natural prior distribution for parameters that could be considered as sums of exponential variables.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # lets simulate
                    a <- rgamma(1000,shape=4,rate=4)
                    # we expect a mean of 1
                    mean(a)
                    
                    # create a random variable
                    x ~ dnGamma(shape=4,rate=1)
                    x
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnExponential'); return false;">dnExponential</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnInvWishart">
    <span onClick="navigateHelp('dnInvWishart');">
        <a href=""><h2>dnInvWishart</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is the constructor of the inverse-Wishart distribution. The inverse-Wishart distribution is defined over positive-definite real symmetric matrices. It can be used as the conjugate prior for the covariance matrix of a multivariate normal model or in the context of a multivariate phylogenetic Brownian model (see dnBrownianMultivariate and example below).
            </p></div>    
    <a href="" onClick="showMore('dnInvWishart'); return false;" class="indent more-label" id="dnInvWishart-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnInvWishart'); return false;" class="indent more-label" id="dnInvWishart-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnInvWishart-more-content"><br />
<h2>Name</h2>
<p class="indent">dnInvWishart</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is the constructor of the inverse-Wishart distribution. The inverse-Wishart distribution is defined over positive-definite real symmetric matrices. It can be used as the conjugate prior for the covariance matrix of a multivariate normal model or in the context of a multivariate phylogenetic Brownian model (see dnBrownianMultivariate and example below).
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnInvWishart( RealSymmetricMatrix sigma0 = [ [ 0.0000 ] ] , RealPos[] diagonal = [  ], RealPos kappa = 0, Natural dim = 0, Natural df )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >sigma0</b>
<div class="indent">
The scale matrix: an arbitrary positive-definite real symmetric matrix.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealSymmetricMatrix<br />
<span class="argument">Default value</span>
[ [ 0.0000 ] ] <br />
</div>
<br />
<b >diagonal</b>
<div class="indent">
A vector of positive real numbers. If this parameter is specified, the scale matrix will be diagonal, with eigenvalues given by this vector.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos[]<br />
<span class="argument">Default value</span>
[  ]<br />
</div>
<br />
<b >df</b>
<div class="indent">
The number of degrees of freedom (should be at least equal to dim-1).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
<b >kappa</b>
<div class="indent">
If kappa is specified, together with the order (dim), the scale matrix parameter of the inverse-Wishart distribution will be equal to kappa*I, where I is the dim x dim identity matrix.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
0<br />
</div>
<br />
<b >dim</b>
<div class="indent">
The order of the matrix. This parameter should only be used in combination with kappa. In other cases, the order of the matrix will be inferred directly from the vector (diagonal) or the scale matrix (sigma0).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Natural<br />
<span class="argument">Default value</span>
0<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            Mathematically, the inverse-Wishart of order dim (defined on dim x dim symmetric matrices) has two parameters: a scale matrix (sigma0) and the number of degrees of freedom (df>dim-1). If Sigma is inverse-Wishart of parameter sigma0 and with df degrees of freedom, then Sigma^(-1) is Wishart of parameter sigma0^(-1) and df degrees of freedom (see dnWishart).
            </p>
<p class="indent">
                This implementation of the inverse Wishart distribution offers three alternative specializations for defining the scale matrix sigma0:
                </p>
<p class="indent">
                - an arbitrary positive-definite symmetrix real matrix M: Sigma ~ dnInvWishart(sigma0=M, df=n).
                </p>
<p class="indent">
                - a positive-definite diagonal matrix (entirely specified by a vector V of positive real numbers): Sigma ~ dnInvWishart(diagonal=V, df=n). 
                </p>
<p class="indent">
                - a multiple of the identity matrix (entirely specified by a positive real number k and the dimension of the matrix d): Sigma ~ dnInvWishart(kappa=k, dim=d, df=n).
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # load a tree
                 trueTree = readTrees("data/chronoplac40.tree")[1]

                 # load quantitative trait data
                 contData = readContinuousCharacterData("data/plac40lhtlog.nex")
                 # get number of traits (dimension of multivariate Brownian process)
                 nTraits <- contData.nchar()[1]

                 # use an inverse-Wishart prior distribution for the rate of the Brownian process
                 # of parameters kappa * Identity and df = nTraits+2 degrees of freedom
                 # where kappa is itself a scaling factor
                 kappa ~ dnLogUniform(min=0.001,max=1000)
                 # can be useful to start from a reasonable value for kappa
                 kappa.setValue(0.1)
                 Sigma ~ dnInvWishart(dim=nTraits, kappa=kappa, df=nTraits+2)

                 # multivariate Brownian process along the tree
                 X ~ dnBrownianMultiVariate(trueTree,sigma=Sigma)

                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Wishart, J. (1928). Biometrika 20A (1-2): 32-52.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnLogUniform">
    <span onClick="navigateHelp('dnLogUniform');">
        <a href=""><h2>dnLogUniform</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            A strictly positive real number x has a log-uniform distribution over interval (min,max) if its logarithm y = ln(x) has uniform distribution over interval (ln(min),ln(max)).
            </p></div>    
    <a href="" onClick="showMore('dnLogUniform'); return false;" class="indent more-label" id="dnLogUniform-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnLogUniform'); return false;" class="indent more-label" id="dnLogUniform-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnLogUniform-more-content"><br />
<h2>Name</h2>
<p class="indent">dnLogUniform</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            A strictly positive real number x has a log-uniform distribution over interval (min,max) if its logarithm y = ln(x) has uniform distribution over interval (ln(min),ln(max)).
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnLogUniform( RealPos min, RealPos max )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >min</b>
<div class="indent">
The lower bound of the interval over which the variable is defined.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >max</b>
<div class="indent">
The upper bound of the interval over which the variable is defined.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            The log-uniform distribution is defined over strictly positive real numbers. Saying that x is log-uniform is equivalent to saying that y = ln(x) is uniform. The log-uniform distribution therefore expresses lack of information about the order of magnitude of a scale parameter: if x has a log-uniform distribution, then it has equal chance to be contained by any of the intervals of the form (10^k, 10^(k+1)) within the allowed range.
                </p>
<p class="indent">
            The density is p(x) = 1/x, which can be seen by defining x = exp(y) where y has uniform distribution and apply the change-of-variable formula.
                </p>
<p class="indent">
            The log-uniform distribution is improper when defined over the entire positive real line. To always make it proper, in RevBayes, a min and a max should always be specified.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # a log-uniform prior over the rate of change of a Brownian trait (or a Brownian relaxed clock)
                 trueTree = readTrees("data/primates.tree")[1]
                 sigma ~ dnLogUniform(min=0.001, max=1000)
                 X ~ dnBrownian(trueTree,sigma) 
                 ...
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent"><br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnUniform'); return false;">dnUniform</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnMixture">
    <span onClick="navigateHelp('dnMixture');">
        <a href=""><h2>dnMixture</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The "dnMixture" distribution specifies that a random variable is distributed according to a mixture distribution.
            </p></div>    
    <a href="" onClick="showMore('dnMixture'); return false;" class="indent more-label" id="dnMixture-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnMixture'); return false;" class="indent more-label" id="dnMixture-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnMixture-more-content"><br />
<h2>Name</h2>
<p class="indent">dnMixture</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The "dnMixture" distribution specifies that a random variable is distributed according to a mixture distribution.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnMixture( RateMatrix[]|Real[]|RealPos[]|Natural[]|Integer[]|Probability[] values, Simplex probabilities )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >values</b>
<div class="indent">
Mixture components. The vector of values according to which the random variable is distributed.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RateMatrix[]|Real[]|RealPos[]|Natural[]|Integer[]|Probability[]<br />
</div>
<br />
<b >probabilities</b>
<div class="indent">
Mixture weights. The probabilities that the random variable is distributed according to each mixture component.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                Let's assume a random variable X is specified as coming from a mixture distribution with mixture components "values" and mixture weights "probabilities". To simulate values for X according to this mixture distribution, we need to follow these steps: (1) with probability probabilities[i] an index i is chosen. (2) Then the value found in values[i] is retrieved. (3) Then this value is assigned to X.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # We simulate a variable coming from a mixture distribution.
                 # Here we use a mixture distribution on 2 normal distributions.
                 stdev <- 1.0
                 mean[1] <- 0.0
                 mean[2] <- 10
                 #We create the vector of Normal distributions.
                 values[1] ~ dnNormal(mean[1], stdev)
                 values[2] ~ dnNormal(mean[2], stdev)
                 
                 print(values[1])
                 print(values[2])

                 #We create a simplex
                 probaOfFirst ~ dnBeta (1.0, 1.0)
                 probabilities <- simplex(probaOfFirst, 1.0)
				 
				 print(probabilities)
				 
                 #Now we can create our mixture distribution:
                 X ~ dnMixture(values, probabilities)
                 
                 # X will take the value values[1] with probability probabilities[1], and the value values[2] with probability probabilities[2].
                 print (X)
                 
                 # We can test that this indeed happens by repeating the drawing of X above with a loop.
				 # We do 1000 replicates:
				 nrep<-1000
				 clear(X)
				 for (i in 1:nrep) {
					X[i] ~ dnMixture(values, probabilities)
				 }

				 # And then counting:
				 num1 = 0
				 for (i in 1:nrep) {
					if (X[i] == values[1] ) {
						num1 = num1 + 1
					}
				 }

				 print ("Proportion of variables drawn with value values[1]: " + (num1/nrep))
				 print ("Proportion of variables drawn with value values[2]: " + ((nrep-num1)/nrep))
                 
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Bastien Boussau</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnDPP'); return false;">dnDPP</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnNormal">
    <span onClick="navigateHelp('dnNormal');">
        <a href=""><h2>dnNormal</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Normal (gaussian) distribution with mean equal to ‘mean’ and standard deviation equal to ‘sd’.
            </p></div>    
    <a href="" onClick="showMore('dnNormal'); return false;" class="indent more-label" id="dnNormal-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnNormal'); return false;" class="indent more-label" id="dnNormal-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnNormal-more-content"><br />
<h2>Name</h2>
<p class="indent">dnNormal</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Normal (gaussian) distribution with mean equal to ‘mean’ and standard deviation equal to ‘sd’.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnNormal( Real mean = 0, RealPos sd = 1 )
                 dnNorm( Real mean = 0, RealPos sd = 1 )
                 dnormal(Real x, Real mean, RealPos sd, Bool log)
                 dnorm(Real x, Real mean, RealPos sd, Bool log)
                 pnormal(Real x, Real mean, RealPos sd)
                 pnorm(Real x, Real mean, RealPos sd, Bool log)
                 qnormal(Probability p, Real mean, RealPos sd)
                 qnorm(Probability p, Real mean, RealPos sd)
                 rnormal(Natural n, Real mean, RealPos sd)
                 rnorm(Natural n, Real mean, RealPos sd)
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >mean</b>
<div class="indent">
The mean of the normal distribution<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
<span class="argument">Default value</span>
0<br />
</div>
<br />
<b >sd</b>
<div class="indent">
The standard deviation.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >x</b>
<div class="indent">
An observation.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
<b >p</b>
<div class="indent">
The probability of the quantile.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Probability<br />
</div>
<br />
<b >n</b>
<div class="indent">
Number of random draws.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                ‘dnorm’ gives the density, ‘pnorm’ gives the distribution function, ‘qnorm’ gives the quantile function, 
                and ‘rNormal’ generates random deviates. The length of the result is determined by ‘n’ for ‘rnorm’.
                </p>
<p class="indent">
                The normal distribution has density
                
                   f(x) = 1/(sqrt(2 pi) sigma) e^-((x - mu)^2/(2 sigma^2))     
                   
                where mu is the mean of the distribution and sigma the standard deviation.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # we simulate some oversations
                 x <- rnorm(n=10,mean=5,sd=10)
                 # let's see what the minum is (you could do the max too)
                 min(x)
                 # let's also see what the mean and the variance are
                 mean(x)
                 var(x)
                 sd(x)
                 # the probability density of the smalles value
                 dNormal
                 # we can compute the probability density of the simulated variables
                 trueTree <- readTrees("data/primates.tree")[1]
                 ...
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnExponential'); return false;">dnExponential</a><br />
<a href="" onClick="navigateHelp('dnLognormal'); return false;">dnLognormal</a><br />
<a href="" onClick="navigateHelp('dnGamma'); return false;">dnGamma</a><br />
<a href="" onClick="navigateHelp('dnUniform'); return false;">dnUniform</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnOUP">
    <span onClick="navigateHelp('dnOUP');">
        <a href=""><h2>dnOUP</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is the constructor function for the Ornstein-Uhlenbeck process of quantitative trait evolution along a time-calibrated phylogeny (Lande, 1976, Felsenstein, 1988). 
            </p></div>    
    <a href="" onClick="showMore('dnOUP'); return false;" class="indent more-label" id="dnOUP-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnOUP'); return false;" class="indent more-label" id="dnOUP-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnOUP-more-content"><br />
<h2>Name</h2>
<p class="indent">dnOrnsteinUhlenbeck</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is the constructor function for the Ornstein-Uhlenbeck process of quantitative trait evolution along a time-calibrated phylogeny (Lande, 1976, Felsenstein, 1988). 
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnOrnsteinUhlenbeck( TimeTree tree, RealPos sigma, Real mean, RealPos phi )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >tree</b>
<div class="indent">
A time-calibrated phylogeny.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
TimeTree<br />
</div>
<br />
<b >sigma</b>
<div class="indent">
The evolutionary rate parameter (variance per unit of time).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >mean</b>
<div class="indent">
The stationary mean of the process.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
<b >phi</b>
<div class="indent">
The damping coefficient (rate at which the process returns to equilbrium).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The Ornstein-Uhlenbeck process is a simple stationary, Gaussian and Markovian univariate process. It can be seen as a combination of a deterministic force (of strength phi) pulling the trait back toward an equilibrium value (the stationary mean of the process) and a Gaussian white-noise perturbation, constantly and randomly pushing the trait away from its equilibrium. The process has stationary variance given by sigma/(2*phi). The typical time for the process to return to equilbirium is given by 1/phi.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # load a tree
                 trueTree = readTrees("data/chronoplac40.tree")[1]

                 # put a gamma distribution on the evolutionary rate parameter 
                 sigma ~ dnGamma(shape=1, scale=1)

                 # put a gamma distribution on the damping rate 
                 phi~ dnGamma(shape=1, scale=1)

                 # put a uniform distribution on the stationary mean
                 mean ~ dnNormal(mean=0, var=10)

                 # create an Ornstein Uhlenbeck process
                 X ~ dnOrnsteinUhlenbeck(tree=trueTree, sigma=sigma, mean=mean, phi=phi) 

                 # load quantitative trait data
                 contData = readContinuousCharacterData("data/plac40lhtlog.nex")

                 # clamp process to second column of quantitative trait data (log body mass in that case)
                 X.clampAt(contData,2)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Lande R, 1976. Natural selection and random genetic drift in phenotypic evolution. Evolution. 30:314-334.
                Felsenstein J, 1988. Phylogenies and quantitative characters. Ann. Rev. Ecol. Syst. 19:445-471.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Your name here</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnOrnsteinUhlenbeck">
    <span onClick="navigateHelp('dnOrnsteinUhlenbeck');">
        <a href=""><h2>dnOrnsteinUhlenbeck</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is the constructor function for the Ornstein-Uhlenbeck process of quantitative trait evolution along a time-calibrated phylogeny (Lande, 1976, Felsenstein, 1988). 
            </p></div>    
    <a href="" onClick="showMore('dnOrnsteinUhlenbeck'); return false;" class="indent more-label" id="dnOrnsteinUhlenbeck-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnOrnsteinUhlenbeck'); return false;" class="indent more-label" id="dnOrnsteinUhlenbeck-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnOrnsteinUhlenbeck-more-content"><br />
<h2>Name</h2>
<p class="indent">dnOrnsteinUhlenbeck</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is the constructor function for the Ornstein-Uhlenbeck process of quantitative trait evolution along a time-calibrated phylogeny (Lande, 1976, Felsenstein, 1988). 
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnOrnsteinUhlenbeck( TimeTree tree, RealPos sigma, Real mean, RealPos phi )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >tree</b>
<div class="indent">
A time-calibrated phylogeny.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
TimeTree<br />
</div>
<br />
<b >sigma</b>
<div class="indent">
The evolutionary rate parameter (variance per unit of time).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >mean</b>
<div class="indent">
The stationary mean of the process.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Real<br />
</div>
<br />
<b >phi</b>
<div class="indent">
The damping coefficient (rate at which the process returns to equilbrium).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The Ornstein-Uhlenbeck process is a simple stationary, Gaussian and Markovian univariate process. It can be seen as a combination of a deterministic force (of strength phi) pulling the trait back toward an equilibrium value (the stationary mean of the process) and a Gaussian white-noise perturbation, constantly and randomly pushing the trait away from its equilibrium. The process has stationary variance given by sigma/(2*phi). The typical time for the process to return to equilbirium is given by 1/phi.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # load a tree
                 trueTree = readTrees("data/chronoplac40.tree")[1]

                 # put a gamma distribution on the evolutionary rate parameter 
                 sigma ~ dnGamma(shape=1, scale=1)

                 # put a gamma distribution on the damping rate 
                 phi~ dnGamma(shape=1, scale=1)

                 # put a uniform distribution on the stationary mean
                 mean ~ dnNormal(mean=0, var=10)

                 # create an Ornstein Uhlenbeck process
                 X ~ dnOrnsteinUhlenbeck(tree=trueTree, sigma=sigma, mean=mean, phi=phi) 

                 # load quantitative trait data
                 contData = readContinuousCharacterData("data/plac40lhtlog.nex")

                 # clamp process to second column of quantitative trait data (log body mass in that case)
                 X.clampAt(contData,2)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Lande R, 1976. Natural selection and random genetic drift in phenotypic evolution. Evolution. 30:314-334.
                Felsenstein J, 1988. Phylogenies and quantitative characters. Ann. Rev. Ecol. Syst. 19:445-471.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Your name here</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnPhyloCTMC">
    <span onClick="navigateHelp('dnPhyloCTMC');">
        <a href=""><h2>dnPhyloCTMC</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            A continuous time Markov model along a given phylogeny.
            </p></div>    
    <a href="" onClick="showMore('dnPhyloCTMC'); return false;" class="indent more-label" id="dnPhyloCTMC-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnPhyloCTMC'); return false;" class="indent more-label" id="dnPhyloCTMC-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnPhyloCTMC-more-content"><br />
<h2>Name</h2>
<p class="indent">dnPhyloCTMC</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            A continuous time Markov model along a given phylogeny.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnPhyloCTMC( Tree tree, RateMatrix Q, Simplex rootFrequencies = NULL, RealPos branchRates = 1, RealPos[] siteRates = [  ], Probability pInv = 0, Natural nSites = 10, String type = DNA )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >tree</b>
<div class="indent">
A description of the argument.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Tree<br />
</div>
<br />
<b >Q</b>
<div class="indent">
Either a global instantaneous rate matrix or a vector of branch-wise rate matrices.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RateMatrix or RateMatrix[]<br />
</div>
<br />
<b >rootFrequencies</b>
<div class="indent">
The frequencies of the states at root. If not provided, this will be computed from the rate matrix<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
<span class="argument">Default value</span>
NULL<br />
</div>
<br />
<b >branchRates</b>
<div class="indent">
A global rate multiplier (branch or clock rate). This can also be per branch rates.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos or RealPos[]<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >siteRates</b>
<div class="indent">
The per site rates for the rate variation among sites.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos[]<br />
<span class="argument">Default value</span>
[ 1 ]<br />
</div>
<br />
<b >pInv</b>
<div class="indent">
The probability of a site being invariant.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Probability<br />
<span class="argument">Default value</span>
0<br />
</div>
<br />
<b >nSites</b>
<div class="indent">
Number of sites used for simulation and initialization.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
<span class="argument">Default value</span>
10<br />
</div>
<br />
<b >type</b>
<div class="indent">
What type of character do we use.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
<span class="argument">Default value</span>
DNA<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The dnPhyloCTMC is the central probability distribution in most phylogenetic analyses.
                The dnPhyloCTMC is a branch-wise continuous time Markov process that models the evolution of a character along a branch.
                This specific implementation uses the sum-product (or pruning) algorithm to integrate over the state at the interior nodes.
                </p>
<p class="indent">
                    The dnPhyloCTMC is specified by the instantaneous rate matrix Q. You can provide either a global rate matrix or a vector of rate matrices.
                    If you provide a vector of rate matrices then the number of rates matrices needs to be equal to the number of branches.
                    However, you may use the same rate matric multiple times in the vector.
                    Note that if you use a branch-heterogeneous substitution process then you must provide a vector of root frequencies.
                    Additionally, if you want to model a time-heterogeneous substitution process, then you only need to provide a time-heterogeneous rate matrix.
                    This will be useful for epoch models.
                </p>
<p class="indent">
                    Generally speaking, the rate of evolution is proportional to the branch length. To exactly get the rate of evolution we scale the branch lengths
                    by a rate-multiplier wich is composed off two facters: (1) the branch rate and (2) the site rate.
                    If you want to use a global clock then you should specify the branchRate parameter.
                    If you want to use a relaxed clock then you need to provide a vector of branch-wise rates, which again needs to be of the size equal to the number of branches.
                    If you want to have a different substitution rate for different partitions then you can use the branchRate parameter too.
                </p>
<p class="indent">
                    Very often you want to specify among site rate variation. Here you can provide any vector of per site rate multiplier.
                    However, you should be aware that the model is only identyfiable if the mean rate is equal to 1.
                    Most commonly you will use 4 quantiles of a gamma distribution but you can use any normalized rates vector.
                    Here, we integrate over all possible assignments of a site being in either rate.
                    Additionally, you can specify the probability of a site being invariable which is the same as having another rate
                    with value 0.0.
                </p>
<p class="indent">
                    In RevBayes we always initialize values with a random draw. In order to do so we need to know the number of sites this data matrix should have and the data type.
                    We cannot infer the data type from the rate matrix because, e.g., DNA, RNA and morphological characters could all have 4 possible states.
                    The initialization actually gives you a nice way to simulate data!
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # read an example tree
                 mytree <- readTrees("data/primates.tree")[1]
                 rateMatrix <- fnJC(4)
                 seq ~ dnPhyloCTMC(tree=myTree, Q=rateMatrix, branchRates=1, nSites=1000, type="DNA")
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('fnGTR'); return false;">fnGTR</a><br />
<a href="" onClick="navigateHelp('fnJC'); return false;">fnJC</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnUniform">
    <span onClick="navigateHelp('dnUniform');">
        <a href=""><h2>dnUniform</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Uniform probabability density.
            </p></div>    
    <a href="" onClick="showMore('dnUniform'); return false;" class="indent more-label" id="dnUniform-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnUniform'); return false;" class="indent more-label" id="dnUniform-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnUniform-more-content"><br />
<h2>Name</h2>
<p class="indent">dnUniform</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Uniform probabability density.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnUniform( RealPos lower = 0, RealPos upper = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >lower</b>
<div class="indent">
The lower bound.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
0<br />
</div>
<br />
<b >upper</b>
<div class="indent">
The upper bound.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The uniform probability density is given by
                f(x) = 1 / (b-a)
                if x is within a and b and 0 otherwise.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # draw some random numbers
                    a <- runif(10,0,1)
                    # compute the max
                    max(a)
                    
                    # now let us create a random variable
                    x ~ dnUniform(-10,10)
                    # note that the type is a real number
                    type(x)
                    
                    # now let us create another random variable
                    y ~ dnUniform(0,10)
                    # note that the type is a positive real number
                    type(y)
                    
                    # now let us create yet another random variable
                    z ~ dnUniform(0,1)
                    # note that the type is a probability because it is between 0 and 1
                    type(z)
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('dnNormal'); return false;">dnNormal</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnUniformTopology">
    <span onClick="navigateHelp('dnUniformTopology');">
        <a href=""><h2>dnUniformTopology</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Specifies equal probability for each possible tree topology with n taxa.
            </p></div>    
    <a href="" onClick="showMore('dnUniformTopology'); return false;" class="indent more-label" id="dnUniformTopology-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnUniformTopology'); return false;" class="indent more-label" id="dnUniformTopology-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnUniformTopology-more-content"><br />
<h2>Name</h2>
<p class="indent">dnUniformTopology</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Specifies equal probability for each possible tree topology with n taxa.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnUniformTopology( String[] names )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >names</b>
<div class="indent">
The taxon names that will be used as the tip names.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String[]<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This simply specifies a uniform probability distribution on all possible tree topologies with n taxa.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 names <- ["A", "B", "C", "D"]
                 topology ~ dnUniformTopology( names )
                 topology
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('treeAssembly'); return false;">treeAssembly</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="dnWishart">
    <span onClick="navigateHelp('dnWishart');">
        <a href=""><h2>dnWishart</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="distribution" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is the constructor of the Wishart distribution. This distribution is defined over positive-definite real symmetric matrices. It can be used as a prior for covariance matrices, although the inverse-Wishart distribution is mathematically more convenient for that purpose (see dnInvWishart).
            </p></div>    
    <a href="" onClick="showMore('dnWishart'); return false;" class="indent more-label" id="dnWishart-more-label" >Show more... >></a>
    <a href="" onClick="showLess('dnWishart'); return false;" class="indent more-label" id="dnWishart-less-label" ><< Show less</a>
    <div class="indent more-content" id="dnWishart-more-content"><br />
<h2>Name</h2>
<p class="indent">dnWishart</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is the constructor of the Wishart distribution. This distribution is defined over positive-definite real symmetric matrices. It can be used as a prior for covariance matrices, although the inverse-Wishart distribution is mathematically more convenient for that purpose (see dnInvWishart).
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 dnWishart( Natural df, RealPos kappa, Natural dim )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >df</b>
<div class="indent">
The number of degrees of freedom (should be at least equal to dim-1).<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
<b >kappa</b>
<div class="indent">
The scaling factor.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >dim</b>
<div class="indent">
The order of the matrix.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
            The Wishart distribution is defined over positive-definite real symmetric matrices. Mathematically, it has two parameters: a dim x dim symmetric definite positive real matrix sigma0 (the scale matrix) and an integer df>dim-1 (the number of degrees of freedom). A random draw from a Wishart can be obtained by drawing dim iid multivariate normal random variates and calculating their scatter matrix (sum-of-squares): Sigma = sum_i X_i X_i'.
            </p>
<p class="indent">
            The present implementation of the Wishart distribution only allows for a matrix parameter sigma0 that is equal to a multiple of the identity matrix (entirely specified by a positive real number kappa and the dimension of the matrix dim): Sigma ~ dnWishart(kappa=k, dim=d, df=n).
            </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 # a Wishart distribution of parameters k * I_2 and df = 4 degrees of freedom
                 # where k is itself a scaling factor endowed with a log-uniform distribution:
                 k ~ dnLogUniform(min=0.001,max=1000)
                 Sigma ~ dnWishart(dim=2, kappa=k, df=4)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Wishart, J. (1928). Biometrika 20A (1-2): 32-52.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

</div>
            
            <br/>
            <br/>
            <div class="section-label"><strong><em>Moves</em></strong></div>            
            <div id="moveEntries"><div id="mvConjugateInverseWishartBrownian">
    <span onClick="navigateHelp('mvConjugateInverseWishartBrownian');">
        <a href=""><h2>mvConjugateInverseWishartBrownian</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This move, described in Lartillot and Poujol, 2011, relies on the conjugate relation between the inverse-Wishart prior and the multivariate normal model to propose a more efficient update of the covariance matrix of a Brownian multivariate model of trait evolution along the phylogeny.
            </p></div>    
    <a href="" onClick="showMore('mvConjugateInverseWishartBrownian'); return false;" class="indent more-label" id="mvConjugateInverseWishartBrownian-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvConjugateInverseWishartBrownian'); return false;" class="indent more-label" id="mvConjugateInverseWishartBrownian-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvConjugateInverseWishartBrownian-more-content"><br />
<h2>Name</h2>
<p class="indent">mvConjugateInverseWishartBrownian</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This move, described in Lartillot and Poujol, 2011, relies on the conjugate relation between the inverse-Wishart prior and the multivariate normal model to propose a more efficient update of the covariance matrix of a Brownian multivariate model of trait evolution along the phylogeny.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvConjugateInverseWishartBrownian( RealSymmetricMatrix sigma, MultivariateRealNodeValTree process, RealPos kappa, Natural df, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >sigma</b>
<div class="indent">
The covariance matrix (should be from an inverse-Wishart distribution, of scale matrix kappa*I and df degrees of freedom).<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
RealSymmetricMatrix<br />
</div>
<br />
<b >process</b>
<div class="indent">
The multivariate process of trait evolution (should be a multivariate Brownian process).<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
MultivariateRealNodeValTree<br />
</div>
<br />
<b >kappa</b>
<div class="indent">
The scaling factor of the inverse-Wishart distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >df</b>
<div class="indent">
The number of degrees of freedom of the inverse-Wishart distribution.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
<b >weight</b>
<div class="indent">
The expected number of times this move will be applied per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This move relies on the specific conjugate relation existing between the inverse-Wishart and the multivariate normal distribution. In the present case, it is more specifically applied to the multivariate Brownian model of trait evolution. The covariance matrix of the Brownian process should be more specifically parameterized as follows: Sigma ~ invWishart(kappa*I, df), i.e. it should have scale matrix proportional to the identity matrix (I), with proportionality factor kappa called the scaling factor; df is the number of degrees of freedom of the inverse-Wishart (see also dnInvWishart). For more details on how this Gibbs sampling strategy works, see Lartillot and Poujol, 2011.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >

		# load trait data 
		contData <- readContinuousCharacterData("data/plac40lhtlog.nex")
		nTraits <- contData.nchar()[1]

		# load tree
		tree <- readTrees("data/chronoplac40.tree")[1]

		# covariance matrix with free scaling factor kappa
		kappa ~ dnLogUniform(min=0.001,max=1000)
		Sigma ~ dnInvWishart(dim=nTraits, kappa=kappa, df=nTraits+1)

		# bivariate Brownian process along the tree
		X ~ dnBrownianMultiVariate(tree,sigma=Sigma)

		# clamp process at the tips of the tree to observed trait values
		for (i in 1:nTraits)	{
			X.clampAt(contData,i,i)
		}

		... 

		# here is a complete set of moves for kappa, sigma and X:

		# set up index for array of moves
		index <- 1

		# moves on kappa
		moves[index++] = mvScale(kappa, lambda=2.0, tune=true, weight=3.0)

		# conjugate Gibbs move on covariance matrix
		moves[index++] = mvConjugateInverseWishartBrownian(sigma=Sigma, process=X, kappa=kappa, df=df, weight=1) 

		# moves on the Brownian process
		moves[index++] = mvMultivariatePhyloProcessSliding(process=X,lambda=1,tune=true,weight=100) 
		moves[index++] = mvMultivariatePhyloProcessTranslation(process=X,lambda=0.1,tune=true,weight=1) 

		# this can be combined with other moves, depending on all other aspects of the model...
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                Lartillot N and Poujol R. 2011. A phylogenetic model for investigating correlated evolution of substitution rates and continuous phenotypic characters. Mol Biol Evol 28:729-744.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mvDPPAllocateAuxGibbs">
    <span onClick="navigateHelp('mvDPPAllocateAuxGibbs');">
        <a href=""><h2>mvDPPAllocateAuxGibbs</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This move re-assigns the data elements to new parameter categories, when those elements are assigned a 
            Dirichlet process prior (DPP). The algorithm for this update is described in Neal (2000; Algorithm 8).
            </p></div>    
    <a href="" onClick="showMore('mvDPPAllocateAuxGibbs'); return false;" class="indent more-label" id="mvDPPAllocateAuxGibbs-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvDPPAllocateAuxGibbs'); return false;" class="indent more-label" id="mvDPPAllocateAuxGibbs-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvDPPAllocateAuxGibbs-more-content"><br />
<h2>Name</h2>
<p class="indent">mvDPPAllocateAuxGibbs</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This move re-assigns the data elements to new parameter categories, when those elements are assigned a 
            Dirichlet process prior (DPP). The algorithm for this update is described in Neal (2000; Algorithm 8).
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvDPPAllocateAuxGibbs( Any[] x, Integer numAux = 4, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The vector of values distributed under the DPP. (See dnDPP for allowed value types)<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
Any[]<br />
</div>
<br />
<b >numAux</b>
<div class="indent">
The number of auxiliary categories proposed by the algorithm.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Integer<br />
<span class="argument">Default value</span>
4<br />
</div>
<br />
<b >weight</b>
<div class="indent">
The expected number of times this move will be applied per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This move is described by Neal (2000) as Algorithm 8 -- "Gibbs Sampling with Auxiliary Parameters". 
                Here, the parameter-category assignment of data elements under the DPP are updated by  
                removing an element from its current category, then reassigning it to each of the remaining categories
                plus numAux additional "auxiliary" categories. This is done for each data element sequentially.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
	 ## number of DPP elements
	 N <- 10
 
	 ## prior mean number of categories
	 k <- 8
 
	 ## concentration parameter of the DPP
	 a <- fnDPPConcFromMean(k, N)
 
	 ## G_0 distribution
	 G_0 = dnExponential(1.0)
 
	 ## DPP distribution constructior
	 rates ~ dnDPP(a, G_0, N)
	 
	 ## Iterator for moves vector
	 mi = 1
	 
	 ## Add mvDPPScaleCatVals to moves vector
	 moves[mi++] = mvDPPScaleCatVals(rates, 0.05, 3.0)

	 ## Add mvDPPAllocateAuxGibbs to moves vector to update parameter-category assignment
	 moves[mi++] = mvDPPAllocateAuxGibbs(rates, 4, 3.0)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
					Heath TA, Holder MT, Huelsenbeck JP. 2012. A Dirichlet process prior for estimating lineage-specific substitution rates. Mol. Biol. Evol. 29(3):939–955. 
            	<br />
<br />
<br />

					Heath TA. 2012. A hierarchical Bayesian model for calibrating estimates of species divergence times. Syst. Biol. 61:793-809.
            	<br />
<br />
<br />

					Huelsenbeck JP, Jain S, Frost SWD, Pond SLK. 2006. A Dirichlet process model for detecting positive selection in protein-coding DNA sequences. Proc Natl Acad Sci U S A. 103:6263–6268.
            	<br />
<br />
<br />

					Huelsenbeck JP, Suchard M. 2007. A nonparametric method for accommodating and testing across-site rate variation. Syst Biol. 56:975–987.
            	<br />
<br />
<br />

					Neal RM. 2000. Markov chain sampling methods for Dirichlet process mixture models. J Comput Graph Stat. 9:249–265.
            	<br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Tracy A. Heath</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mvDPPScaleCatAllocateAux'); return false;">mvDPPScaleCatAllocateAux</a><br />
<a href="" onClick="navigateHelp('mvDPPScaleCatVals'); return false;">mvDPPScaleCatVals</a><br />
<a href="" onClick="navigateHelp('dnDPP'); return false;">dnDPP</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mvDPPScaleCatVals">
    <span onClick="navigateHelp('mvDPPScaleCatVals');">
        <a href=""><h2>mvDPPScaleCatVals</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This scaling move updates the values associated with each parameter category under the DPP, 
            when the variables are RealPos type. This move operates on all of the k parameter categories
            and accepts/rejects within the move, therefore it is always accepted.
            </p></div>    
    <a href="" onClick="showMore('mvDPPScaleCatVals'); return false;" class="indent more-label" id="mvDPPScaleCatVals-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvDPPScaleCatVals'); return false;" class="indent more-label" id="mvDPPScaleCatVals-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvDPPScaleCatVals-more-content"><br />
<h2>Name</h2>
<p class="indent">mvDPPScaleCatVals</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This scaling move updates the values associated with each parameter category under the DPP, 
            when the variables are RealPos type. This move operates on all of the k parameter categories
            and accepts/rejects within the move, therefore it is always accepted.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvDPPScaleCatVals( RealPos[] x, RealPos lambda = 1, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The vector of RealPos values distributed under the DPP.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
RealPos[]<br />
</div>
<br />
<b >lambda</b>
<div class="indent">
The tuning parameter influence the amount of scaling for each category.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >weight</b>
<div class="indent">
The expected number of times this move will be applied per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This move updates the values assigned to a vector of RealPos variables that are assigned a
                Dirichlet process prior (DPP). This simple scale move operates by proposing a new value for each
                of k parameter categories using a multiplier and the prior ratio depends on the base distribution (G_0)
                of the DPP. It is important to note that this move DOES NOT update the parameter category assignment.
                Thus, this move must be coupled with the Gibbs move -- mvDPPAllocateAuxGibbs -- re-assign the 
                data elements to new parameter categories. 
                </p>
<p class="indent">
                See Neal (2000) for descriptions of proposals for DPP-random variables.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
	 ## number of DPP elements
	 N <- 10
 
	 ## prior mean number of categories
	 k <- 8
 
	 ## concentration parameter of the DPP
	 a <- fnDPPConcFromMean(k, N)
 
	 ## G_0 distribution
	 G_0 = dnExponential(1.0)
 
	 ## DPP distribution constructior
	 rates ~ dnDPP(a, G_0, N)
	 
	 ## Iterator for moves vector
	 mi = 1
	 
	 ## Add mvDPPScaleCatVals to moves vector
	 moves[mi++] = mvDPPScaleCatVals(rates, 0.05, 3.0)

	 ## Add mvDPPAllocateAuxGibbs to moves vector to update parameter-category assignment
	 moves[mi++] = mvDPPAllocateAuxGibbs(rates, 4, 3.0)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
					Heath TA, Holder MT, Huelsenbeck JP. 2012. A Dirichlet process prior for estimating lineage-specific substitution rates. Mol. Biol. Evol. 29(3):939–955. 
            	<br />
<br />
<br />

					Heath TA. 2012. A hierarchical Bayesian model for calibrating estimates of species divergence times. Syst. Biol. 61:793-809.
            	<br />
<br />
<br />

					Huelsenbeck JP, Jain S, Frost SWD, Pond SLK. 2006. A Dirichlet process model for detecting positive selection in protein-coding DNA sequences. Proc Natl Acad Sci U S A. 103:6263–6268.
            	<br />
<br />
<br />

					Huelsenbeck JP, Suchard M. 2007. A nonparametric method for accommodating and testing across-site rate variation. Syst Biol. 56:975–987.
            	<br />
<br />
<br />

					Neal RM. 2000. Markov chain sampling methods for Dirichlet process mixture models. J Comput Graph Stat. 9:249–265.
            	<br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Tracy A. Heath</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mvDPPScaleCatAllocateAux'); return false;">mvDPPScaleCatAllocateAux</a><br />
<a href="" onClick="navigateHelp('mvDPPAllocateAuxGibbs'); return false;">mvDPPAllocateAuxGibbs</a><br />
<a href="" onClick="navigateHelp('mvScale'); return false;">mvScale</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mvMatrixSingleElementSliding">
    <span onClick="navigateHelp('mvMatrixSingleElementSliding');">
        <a href=""><h2>mvMatrixSingleElementSliding</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This move function performs a sliding-move on one random entry of a real matrix.
            </p></div>    
    <a href="" onClick="showMore('mvMatrixSingleElementSliding'); return false;" class="indent more-label" id="mvMatrixSingleElementSliding-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvMatrixSingleElementSliding'); return false;" class="indent more-label" id="mvMatrixSingleElementSliding-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvMatrixSingleElementSliding-more-content"><br />
<h2>Name</h2>
<p class="indent">mvMatrixSingleElementSliding</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This move function performs a sliding-move on one random entry of a real matrix.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvMatrixSingleElementSliding( RealMatrix x, RealPos lambda = 1, Bool tune = TRUE, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The matrix to which the move is applied.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
RealMatrix<br />
</div>
<br />
<b >lambda</b>
<div class="indent">
The tuning parameter or size of the proposal.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >tune</b>
<div class="indent">
This flag specifies if the proposal tuning parameter (lambda) is retuned during the burnin period.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >weight</b>
<div class="indent">
The expected number of times this move will be applied per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This move proposes a simple, one-component-at-a-time, sliding-move to a real matrix. The sliding move is applied to a randomly chosen entry of the matrix.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 ...
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent"><br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mvNNI">
    <span onClick="navigateHelp('mvNNI');">
        <a href=""><h2>mvNNI</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This is a 'brief' description of the amazing function you have implemented.
            </p></div>    
    <a href="" onClick="showMore('mvNNI'); return false;" class="indent more-label" id="mvNNI-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvNNI'); return false;" class="indent more-label" id="mvNNI-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvNNI-more-content"><br />
<h2>Name</h2>
<p class="indent">mvNNI</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This is a 'brief' description of the amazing function you have implemented.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvNNI( TimeTree|Topology tree, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >tree</b>
<div class="indent">
The tree that we want to change.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
TimeTree|Topology<br />
</div>
<br />
<b >weight</b>
<div class="indent">
How often do we use this move per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                Propose a nearest neighbor interchange to search tree space. That is, we randomly select a node in the tree that is not the root.
                Then we exchange the node with the sibling of its parent. For time trees this proposal will change the node age
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                 names <- ["A", "B", "C", "D"]
                 tau ~ dnUniformTopology( names)
                 moves[1] = mvNNI( tau )
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mvNarrow'); return false;">mvNarrow</a><br />
<a href="" onClick="navigateHelp('mvSPR'); return false;">mvSPR</a><br />
<a href="" onClick="navigateHelp('mvFNPR'); return false;">mvFNPR</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mvScale">
    <span onClick="navigateHelp('mvScale');">
        <a href=""><h2>mvScale</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
                A scaling move draws a random uniform number u ~ unif(-0.5,0.5) and scales the current vale by a scaling factor
                sf = exp( lambda * u )
                where lambda is the tuning parameter of the proposal to influence the size of the proposals.
            </p></div>    
    <a href="" onClick="showMore('mvScale'); return false;" class="indent more-label" id="mvScale-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvScale'); return false;" class="indent more-label" id="mvScale-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvScale-more-content"><br />
<h2>Name</h2>
<p class="indent">mvScale</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                A scaling move draws a random uniform number u ~ unif(-0.5,0.5) and scales the current vale by a scaling factor
                sf = exp( lambda * u )
                where lambda is the tuning parameter of the proposal to influence the size of the proposals.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent">mvScale</p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvScale( RealPos x, RealPos lambda = 1, Bool tune = TRUE, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The random variable that will be scaled<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
RealPos<br />
</div>
<br />
<b >lambda</b>
<div class="indent">
The tuning parameter influence the amount of scaling.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >tune</b>
<div class="indent">
If auto-tuning should be used.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >weight</b>
<div class="indent">
The expected number of times this move will be applied per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel <- model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] <- mScale(a, lambda=1.0, weight=1.0)
                    monitors[1] <- modelmonitor("out.txt")
                    
                    # Create an mcmc object
                    mymcmcObject <- mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.burnin( generations = 400, tuningInterval = 100)
                    mymcmcObject.run( generations = 400)
                    
                    # print the summary of the operators (now tuned)
                    mymcmcObject.operatorSummary()
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mvSlide'); return false;">mvSlide</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mvSimplexElementScale">
    <span onClick="navigateHelp('mvSimplexElementScale');">
        <a href=""><h2>mvSimplexElementScale</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Randomly pick and scale a single element of a simplex.
            </p></div>    
    <a href="" onClick="showMore('mvSimplexElementScale'); return false;" class="indent more-label" id="mvSimplexElementScale-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvSimplexElementScale'); return false;" class="indent more-label" id="mvSimplexElementScale-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvSimplexElementScale-more-content"><br />
<h2>Name</h2>
<p class="indent">mvSimplexElementScale</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Randomly pick and scale a single element of a simplex.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvSimplexElementScale( Simplex x, RealPos alpha = 1, Bool tune = TRUE, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The variable that we want to modify.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
Simplex<br />
</div>
<br />
<b >alpha</b>
<div class="indent">
A concentration how close to the current value our proposals will be.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >tune</b>
<div class="indent">
Should we tune the alpha parameter.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >weight</b>
<div class="indent">
How often we want to use this move per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This move randomly picks an element of the simplex. Then, we use a beta distribution centered on the current value
                with the concentration scaled by the alpha parameter. After proposing a new value we rescale the simplex so that it sums to 1 again.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    a <- [1,1,1,1]
                    b ~ dnDirichlet(a)
                    moves[1] = mvSimplexElementScale(b,alpha=1.0)
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mvSimplex'); return false;">mvSimplex</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mvSymmetricMatrixSimple">
    <span onClick="navigateHelp('mvSymmetricMatrixSimple');">
        <a href=""><h2>mvSymmetricMatrixSimple</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="move" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            This move function performs a sliding-move update on one random entry of a real symmetric matrix.
            </p></div>    
    <a href="" onClick="showMore('mvSymmetricMatrixSimple'); return false;" class="indent more-label" id="mvSymmetricMatrixSimple-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mvSymmetricMatrixSimple'); return false;" class="indent more-label" id="mvSymmetricMatrixSimple-less-label" ><< Show less</a>
    <div class="indent more-content" id="mvSymmetricMatrixSimple-more-content"><br />
<h2>Name</h2>
<p class="indent">mvSymmetricMatrixSimple</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            This move function performs a sliding-move update on one random entry of a real symmetric matrix.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mvSymmetricMatrixSimple( RealSymmetricMatrix x, RealPos lambda = 1, Bool tune = TRUE, RealPos weight = 1 )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >x</b>
<div class="indent">
The matrix to which the move is applied.<br />
<br />
<span class="argument">Argument type: </span>
by-reference<br />
<span class="argument">Value type: </span>
RealSymmetricMatrix<br />
</div>
<br />
<b >lambda</b>
<div class="indent">
The tuning parameter or size of the proposal.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >tune</b>
<div class="indent">
This flag specifies if the proposal tuning parameter (lambda) is retuned during the burnin period.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >weight</b>
<div class="indent">
The expected number of times this move will be applied per iteration.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
RealPos<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                This move proposes a simple, one-component-at-a-time, sliding-move update to a real symmetric matrix. The sliding move is applied to a randomly chosen entry of the matrix (and to the symmetric entry if not along the diagonal). 
                </p>
<p class="indent">
                This move function is very general. In the case of covariance matrices, and using an inverse-Wishart prior, another specialized and more efficient Gibbs sampling move can be used instead: see mvConjugateInverseWishartBrownian.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
		# load trait data 
		contData <- readContinuousCharacterData("data/plac40lhtlog.nex")
		nTraits <- contData.nchar()[1]

		# load tree
		tree <- readTrees("data/chronoplac40.tree")[1]

		# covariance matrix with free scaling factor kappa
		kappa ~ dnLogUniform(min=0.001,max=1000)
		Sigma ~ dnInvWishart(dim=nTraits, kappa=kappa, df=nTraits+1)

		# bivariate Brownian process along the tree
		X ~ dnBrownianMultiVariate(tree,sigma=Sigma)

		# clamp process at the tips of the tree to observed trait values
		for (i in 1:nTraits)	{
			X.clampAt(contData,i,i)
		}

		... 

		# set up index for array of moves
		index <- 1

		# a simple move on Sigma
		moves[index++] = mvSymmetricMatrixSimple(Sigma, lambda=1.0, tune=true, weight=10.0)

		# moves on the Brownian process
		moves[index++] = mvMultivariatePhyloProcessSliding(process=X,lambda=1,tune=true,weight=100) 
		moves[index++] = mvMultivariatePhyloProcessTranslation(process=X,lambda=0.1,tune=true,weight=1) 

                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent"><br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Nicolas Lartillot</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('similar function'); return false;">similar function</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

</div>
            
            <br/>
            <br/>
            <div class="section-label"><strong><em>Monitors</em></strong></div>            
            <div id="monitorEntries"><div id="mnFile">
    <span onClick="navigateHelp('mnFile');">
        <a href=""><h2>mnFile</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="monitor" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Monitors variables and prints them into a file.
            </p></div>    
    <a href="" onClick="showMore('mnFile'); return false;" class="indent more-label" id="mnFile-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mnFile'); return false;" class="indent more-label" id="mnFile-less-label" ><< Show less</a>
    <div class="indent more-content" id="mnFile-more-content"><br />
<h2>Name</h2>
<p class="indent">mnFile</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Monitors variables and prints them into a file.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mnFile( RevObject , String filename, Natural printgen = 1, String separator = 	, Bool posterior = TRUE, Bool likelihood = TRUE, Bool prior = TRUE, Bool append = FALSE )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b ></b>
<div class="indent">
This is an ellipsis argument where you can add one or more variables that should be monitored.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RevObject<br />
</div>
<br />
<b >filename</b>
<div class="indent">
The filename including the relative path.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
</div>
<br />
<b >printgen</b>
<div class="indent">
The print frequency how often the variables will be monitor (thinning).<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >separator</b>
<div class="indent">
How should variables/columns be separated. The default is a tab<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
</div>
<br />
<b >posterior</b>
<div class="indent">
Should we monitor the posterior probability as well.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >likelihood</b>
<div class="indent">
Should we monitor the likelihood as well.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >prior</b>
<div class="indent">
Should we monitor the prior probability as well.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >append</b>
<div class="indent">
Should we append to an existing file?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
FALSE<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The implementation of this fancy function ... 
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel = model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] = mvScale(a, lambda=1.0, weight=1.0)
                    monitors[1] = mnFile(a,"output/out.log")
                    
                    # Create an mcmc object
                    mymcmcObject = mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.burnin( generations = 400, tuningInterval = 100)
                    mymcmcObject.run( generations = 400)
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mnModel'); return false;">mnModel</a><br />
<a href="" onClick="navigateHelp('mnScreen'); return false;">mnScreen</a><br />
<a href="" onClick="navigateHelp('Mcmc'); return false;">Mcmc</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mnModel">
    <span onClick="navigateHelp('mnModel');">
        <a href=""><h2>mnModel</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="monitor" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            Monitor all numeric variables of a model and print them to a file.
            </p></div>    
    <a href="" onClick="showMore('mnModel'); return false;" class="indent more-label" id="mnModel-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mnModel'); return false;" class="indent more-label" id="mnModel-less-label" ><< Show less</a>
    <div class="indent more-content" id="mnModel-more-content"><br />
<h2>Name</h2>
<p class="indent">mnModel</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            Monitor all numeric variables of a model and print them to a file.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mnModel( String filename, Natural printgen = 1, String separator = TAB, Bool posterior = TRUE, Bool likelihood = TRUE, Bool prior = TRUE, Bool append = FALSE, Bool stochasticOnly = FALSE )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >filename</b>
<div class="indent">
The file to log to.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
</div>
<br />
<b >printgen</b>
<div class="indent">
The frequency of monitoring (i.e. thinning).<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >separator</b>
<div class="indent">
The separator between variables in the file. The default is a tab.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
</div>
<br />
<b >posterior</b>
<div class="indent">
Should the posterior probability be printed too?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >likelihood</b>
<div class="indent">
Should the likelihood be printed too?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >prior</b>
<div class="indent">
Should the prior probability be printed too?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >append</b>
<div class="indent">
Should we append to an existing file?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
FALSE<br />
</div>
<br />
<b >stochasticOnly</b>
<div class="indent">
Should we monitor stochastic nodes only or deterministic nodes as well?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
FALSE<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The model monitor is very useful to automatically collect all variables in a model and monitors them.
                The model monitor traverses the model graph and collects all numeric variables (e.g., no trees).
                Numeric variables can be read in and interpreted easier than other variables.
                The model monitor prints the current value of the variables every printgen iterations to the specified file.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel = model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] = mvScale(a, lambda=1.0, weight=1.0)
                    monitors[1] = mnModel("output/out.log")
                    
                    # Create an mcmc object
                    mymcmcObject = mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.burnin( generations = 400, tuningInterval = 100)
                    mymcmcObject.run( generations = 400)
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mnFile'); return false;">mnFile</a><br />
<a href="" onClick="navigateHelp('mnScreen'); return false;">mnScreen</a><br />
<a href="" onClick="navigateHelp('mcmc'); return false;">mcmc</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

<div id="mnScreen">
    <span onClick="navigateHelp('mnScreen');">
        <a href=""><h2>mnScreen</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="monitor" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
            The screen monitor prints the values to the screen.
            </p></div>    
    <a href="" onClick="showMore('mnScreen'); return false;" class="indent more-label" id="mnScreen-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mnScreen'); return false;" class="indent more-label" id="mnScreen-less-label" ><< Show less</a>
    <div class="indent more-content" id="mnScreen-more-content"><br />
<h2>Name</h2>
<p class="indent">mnScreen</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
            The screen monitor prints the values to the screen.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent"></p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mnScreen( RevObject , Natural printgen = 1, Bool posterior = TRUE, Bool likelihood = TRUE, Bool prior = TRUE )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b ></b>
<div class="indent">
One or more variables that you want to monitor. This is an ellipsis argument.<br />
<br />
<span class="argument">Argument type: </span>
by-const-reference<br />
<span class="argument">Value type: </span>
RevObject<br />
</div>
<br />
<b >printgen</b>
<div class="indent">
The frequency how often to print to the screen (thinning).<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
<span class="argument">Default value</span>
1<br />
</div>
<br />
<b >posterior</b>
<div class="indent">
Should we monitor the posterior probability too?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >likelihood</b>
<div class="indent">
Should we monitor the likelihood too?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
<b >prior</b>
<div class="indent">
Should we monitor the prior probability too?<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
TRUE<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                The screen monitor prints any variable that you want to the screen.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel = model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] = mvScale(a, lambda=1.0, weight=1.0)
                    monitors[1] = mnScreen(a)
                    
                    # Create an mcmc object
                    mymcmcObject = mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.burnin( generations = 400, tuningInterval = 100)
                    mymcmcObject.run( generations = 400)
                    
                    # print the summary of the operators (now tuned)
                    mymcmcObject.operatorSummary()
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mnModel'); return false;">mnModel</a><br />
<a href="" onClick="navigateHelp('mnFile'); return false;">mnFile</a><br />
<a href="" onClick="navigateHelp('mcmc'); return false;">mcmc</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

</div>
            
            <br/>
            <br/>
            <div class="section-label"><strong><em>Types</em></strong></div>            
            <div id="typeEntries"><div id="mcmc">
    <span onClick="navigateHelp('mcmc');">
        <a href=""><h2>mcmc</h2></a>
    </span> 
    <input type="hidden" id="entry-type" value="type" />
    
    <!--<a href="#entry-file#">View</a>-->
    <div class="indent"><p>
                The MCMC analysis object keeps a model and the associated moves and monitors.
                The object is used to run Markov chain Monte Carlo (MCMC) simulation on
                the model, using the provided moves, to obtain a sample of the posterior probability
                distribution. During the analysis, the monitors are responsible for sampling model parameters of interest.
            </p></div>    
    <a href="" onClick="showMore('mcmc'); return false;" class="indent more-label" id="mcmc-more-label" >Show more... >></a>
    <a href="" onClick="showLess('mcmc'); return false;" class="indent more-label" id="mcmc-less-label" ><< Show less</a>
    <div class="indent more-content" id="mcmc-more-content"><br />
<h2>Name</h2>
<p class="indent">mcmc</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                The MCMC analysis object keeps a model and the associated moves and monitors.
                The object is used to run Markov chain Monte Carlo (MCMC) simulation on
                the model, using the provided moves, to obtain a sample of the posterior probability
                distribution. During the analysis, the monitors are responsible for sampling model parameters of interest.
            </p>
</p>
<h2 >Constructors</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent">mcmc</p>
<h2>Description</h2>
<p class="indent"></p>
<h2>Usage</h2>
<p class="indent">
                 mcmc( Model model, Monitor[] monitors, Move[] moves, String moveschedule = random )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >model</b>
<div class="indent">

                    The model to use in the analysis. The function makes an independent copy of the model
                    and stores it internally in the MCMC object.
                <br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Model<br />
</div>
<br />
<b >monitors</b>
<div class="indent">

                    The monitors to use in the analysis. The function makes an independent copy of the monitors
                    and stores them internally in the MCMC object.
                <br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Monitor[]<br />
</div>
<br />
<b >moves</b>
<div class="indent">

                    The moves to use in the analysis. The function makes an independent copy of the moves
                    and stores them internally in the MCMC object.
                <br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Move[]<br />
</div>
<br />
<b >moveschedule</b>
<div class="indent">

                    The schedule used within this move, which is a mechanism on how to choose new moves.
                    Either a new single move is chosen or a set of moves is chosen proportional to the sum of weights in sequential order or in random order.
                <br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
String<br />
<span class="argument">Default value</span>
random<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                    The MCMC analysis object produced by a call to this function keeps copies of the model and the associated moves and monitors.
                    The MCMC analysis object is used to run Markov chain Monte Carlo (MCMC) simulation on the model, using the provided moves,
                    to obtain a sample of the posterior probability distribution. During the analysis, the monitors are responsible for sampling
                    model parameters of interest.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel = model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] = mvScale(a, lambda=1.0, weight=1.0)
                    monitors[1] = mnFile(a,"output/out.log")
                    
                    # Create an mcmc object
                    mymcmcObject = mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.burnin( generations = 400, tuningInterval = 100)
                    mymcmcObject.run( generations = 400)
                    
                    # print the summary of the operators (now tuned)
                    mymcmcObject.operatorSummary()
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    Metropolis N, AW Rosenbluth, MN Rosenbluth, AH Teller, E Teller (1953) Equation of
                    state calculations by fast computing machines. Journal of Chemical Physics, 21:1087-1092.
                <br />
<br />
<br />

                    Hastings WK (1970) Monte Carlo sampling methods using Markov chains and their applications. Biometrika, 57:97-109.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mcmc.run'); return false;">mcmc.run</a><br />
</p><br />
</div>
<h2 >Methods</h2>
<div class="indent">
<br />
<h2>Name</h2>
<p class="indent">burnin</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                    Run a burnin simulation to let the MCMC procedure find the target distribution and to autotune move tuning parameters.
                    No parameter values are sampled.
                </p>
</p>
<h2>Usage</h2>
<p class="indent">
                 burnin( Natural generations, Natural tuningInterval )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >generations</b>
<div class="indent">
The number of generations of the burnin run.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
<b >tuningInterval</b>
<div class="indent">

                    The tuning interval in number of generations.
                    The tuning interval should not be larger than the number of generations.
                <br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
</div>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel = model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] = mvScale(a, lambda=1.0, weight=1.0)
                    monitors[1] = mnFile(a,"output/out.log")
                    
                    # Create an mcmc object
                    mymcmcObject = mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.burnin( generations = 400, tuningInterval = 100)
                    mymcmcObject.run( generations = 400)
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mcmc.run'); return false;">mcmc.run</a><br />
</p><br />
<br />
<h2>Name</h2>
<p class="indent">operatorSummary</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                    This function provides information about the number of proposals, the acceptance rates and the tuning parameter after an MCMC sampler has run for some iterations.
                </p>
</p>
<h2>Usage</h2>
<p class="indent">
                 operatorSummary(  )
                 </p>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel <- model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] <- mScale(a, lambda=1.0, weight=1.0)
                    monitors[1] <- modelmonitor("out.txt")
                    
                    # Create an mcmc object
                    mymcmcObject <- mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.burnin( generations = 400, tuningInterval = 100)
                    
                    # print the summary of the operators (now tuned)
                    mymcmcObject.operatorSummary()
                 </pre>
</p>
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mcmc.burnin'); return false;">mcmc.burnin</a><br />
<a href="" onClick="navigateHelp('mcmc.run'); return false;">mcmc.run</a><br />
</p><br />
<br />
<h2>Name</h2>
<p class="indent">run</p>
<h2>Description</h2>
<p class="indent"><p class="indent">
                    Running an MCMC analysis for a given number of iterations.
                    This may continue a previous run or start a new run if this is the first run.
                    You can also run the MCMC under the prior which disregards the likelihood contribution from the clamped variables.
                </p>
</p>
<h2>Usage</h2>
<p class="indent">
                 run( Natural generations, Bool underPrior = FALSE )
                 </p>
<h2 >Arguments</h2>
<div class="indent">
<b >generations</b>
<div class="indent">
The number of iterations the MCMC should run.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Natural<br />
</div>
<br />
<b >underPrior</b>
<div class="indent">
If this MCMC should run under the prior by setting all the likelihoods from the clamped variables to 1.0.<br />
<br />
<span class="argument">Argument type: </span>
by-value<br />
<span class="argument">Value type: </span>
Bool<br />
<span class="argument">Default value</span>
FALSE<br />
</div>
<br />
</div>
<h2>Details</h2>
<p class="indent"><p class="indent">
                    Markov chain Monte Carlo (MCMC) simulation is a numerical method that, in this program, is used
                    to approximate the posterior probability distribution of the parameters of a model. The basic
                    sampling mechanism we use is the Metropolis-Hastings algorithm (Metropolis et al. 1953; Hastings 1970).
                </p>
<p class="indent">
                    An MCMC simulation is known to converge to the posterior probability distribution, such that the frequency with
                    which a particular parameter value is sampled is a valid approximation of its posterior probability. However,
                    convergence to the target distribution may be slow, so it is important to assess convergence to ensure that you
                    are likely to have a reasonable sample from the posterior.
                </p>
</p>
<h2>Example</h2>
<p class="indent"><pre >
                    # Create a simple model (unclamped)
                    a ~ exponential(1)
                    mymodel <- model(a)
                    
                    # Create a move vector and a monitor vector
                    moves[1] <- mScale(a, lambda=1.0, weight=1.0)
                    monitors[1] <- modelmonitor("out.txt")
                    
                    # Create an mcmc object
                    mymcmcObject <- mcmc( mymodel, monitors, moves)
                    
                    # Run a short analysis
                    mymcmcObject.run( generations = 100)
                 </pre>
</p>
<h2 >Reference</h2>
<p class="indent">
                    Metropolis N, AW Rosenbluth, MN Rosenbluth, AH Teller, E Teller (1953) Equation of
                    state calculations by fast computing machines. Journal of Chemical Physics, 21:1087-1092.
                <br />
<br />
<br />

                    Hastings WK (1970) Monte Carlo sampling methods using Markov chains and their applications. Biometrika, 57:97-109.
                <br />
<br />
<br />
</p><br />
<h2>Author</h2>
<p class="indent">Sebastian Hoehna</p>
<br />
<h2 >See also</h2>
<p class="indent"><a href="" onClick="navigateHelp('mcmc.burnin'); return false;">mcmc.burnin</a><br />
</p><br />
</div>
<h2>Author</h2>
<p class="indent"></p>
</div>
</div>

</div>
        </div>

        <br />
        <br />
        <br />
        <hr />        
        <p class="footer">
            <b>RevBayes</b>
            <br />
            <i>
                An R-like environment for phylogenetic analysis using Bayesian inference. <br />
                Released under the GPL license, version 3. <br />
            </i>
            Learn more on the <a href="http://sourceforge.net/apps/mediawiki/revbayes/index.php?title=RevBayes_MediaWiki" >RevBayes Wiki</a> 
            <br />
        </p>
    </body>
</html>
